// @gql.tada/cli-utils/dist/chunks/index-chunk.js
var t = require("node:path");

var r = require("@gql.tada/internal");

var i = require("node:fs/promises");

var n = require("node:readline");

var a = require("graphql");

var o = require("node:module");

var s = require("typescript");

var l = require("@0no-co/graphqlsp/api");

function _interopNamespaceDefault(t) {
  var r = Object.create(null);
  if (t) {
    Object.keys(t).forEach((function(i) {
      if ("default" !== i) {
        var n = Object.getOwnPropertyDescriptor(t, i);
        Object.defineProperty(r, i, n.get ? n : {
          enumerable: !0,
          get: function() {
            return t[i];
          }
        });
      }
    }));
  }
  r.default = t;
  return r;
}

var c = _interopNamespaceDefault(t);

var u = _interopNamespaceDefault(i);

var d;

!function(t) {
  t.StartOfInput = "\0";
  t.EndOfInput = "";
  t.EndOfPartialInput = "";
}(d || (d = {}));

var f;

!function(t) {
  t[t.InitialNode = 0] = "InitialNode";
  t[t.SuccessNode = 1] = "SuccessNode";
  t[t.ErrorNode = 2] = "ErrorNode";
  t[t.CustomNode = 3] = "CustomNode";
}(f || (f = {}));

var g = -1;

var p = /^(-h|--help)(?:=([0-9]+))?$/;

var h = /^(--[a-z]+(?:-[a-z]+)*|-[a-zA-Z]+)$/;

var v = /^-[a-zA-Z]{2,}$/;

var m = /^([^=]+)=([\s\S]*)$/;

var y = "1" === process.env.DEBUG_CLI;

class UsageError extends Error {
  constructor(t) {
    super(t);
    this.clipanion = {
      type: "usage"
    };
    this.name = "UsageError";
  }
}

class UnknownSyntaxError extends Error {
  constructor(t, r) {
    super();
    this.input = t;
    this.candidates = r;
    this.clipanion = {
      type: "none"
    };
    this.name = "UnknownSyntaxError";
    if (0 === this.candidates.length) {
      this.message = "Command not found, but we're not sure what's the alternative.";
    } else if (this.candidates.every((t => null !== t.reason && t.reason === r[0].reason))) {
      var [{reason: i}] = this.candidates;
      this.message = `${i}\n\n${this.candidates.map((({usage: t}) => `$ ${t}`)).join("\n")}`;
    } else if (1 === this.candidates.length) {
      var [{usage: n}] = this.candidates;
      this.message = `Command not found; did you mean:\n\n$ ${n}\n${whileRunning(t)}`;
    } else {
      this.message = `Command not found; did you mean one of:\n\n${this.candidates.map((({usage: t}, r) => `${`${r}.`.padStart(4)} ${t}`)).join("\n")}\n\n${whileRunning(t)}`;
    }
  }
}

class AmbiguousSyntaxError extends Error {
  constructor(t, r) {
    super();
    this.input = t;
    this.usages = r;
    this.clipanion = {
      type: "none"
    };
    this.name = "AmbiguousSyntaxError";
    this.message = `Cannot find which to pick amongst the following alternatives:\n\n${this.usages.map(((t, r) => `${`${r}.`.padStart(4)} ${t}`)).join("\n")}\n\n${whileRunning(t)}`;
  }
}

var whileRunning = t => `While running ${t.filter((t => t !== d.EndOfInput && t !== d.EndOfPartialInput)).map((t => {
  var r = JSON.stringify(t);
  if (t.match(/\s/) || 0 === t.length || r !== `"${t}"`) {
    return r;
  } else {
    return t;
  }
})).join(" ")}`;

var S = Array(80).fill("━");

for (var w = 0; w <= 24; ++w) {
  S[S.length - w] = `[38;5;${232 + w}m━`;
}

var b = {
  header: t => `[1m━━━ ${t}${t.length < 75 ? ` ${S.slice(t.length + 5).join("")}` : ":"}[0m`,
  bold: t => `[1m${t}[22m`,
  error: t => `[31m[1m${t}[22m[39m`,
  code: t => `[36m${t}[39m`
};

var x = {
  header: t => t,
  bold: t => t,
  error: t => t,
  code: t => t
};

function formatMarkdownish(t, {format: r, paragraphs: i}) {
  t = function dedent(t) {
    var r = t.split("\n");
    var i = r.filter((t => t.match(/\S/)));
    var n = i.length > 0 ? i.reduce(((t, r) => Math.min(t, r.length - r.trimStart().length)), Number.MAX_VALUE) : 0;
    return r.map((t => t.slice(n).trimRight())).join("\n");
  }(t = t.replace(/\r\n?/g, "\n"));
  t = (t = (t = t.replace(/^\n+|\n+$/g, "")).replace(/^(\s*)-([^\n]*?)\n+/gm, "$1-$2\n\n")).replace(/\n(\n)?\n*/g, ((t, r) => r ? r : " "));
  if (i) {
    t = t.split(/\n/).map((t => {
      var r = t.match(/^\s*[*-][\t ]+(.*)/);
      if (!r) {
        return t.match(/(.{1,80})(?: |$)/g).join("\n");
      }
      var i = t.length - t.trimStart().length;
      return r[1].match(new RegExp(`(.{1,${78 - i}})(?: |$)`, "g")).map(((t, r) => " ".repeat(i) + (0 === r ? "- " : "  ") + t)).join("\n");
    })).join("\n\n");
  }
  return (t = (t = t.replace(/(`+)((?:.|[\n])*?)\1/g, ((t, i, n) => r.code(i + n + i)))).replace(/(\*\*)((?:.|[\n])*?)\1/g, ((t, i, n) => r.bold(i + n + i)))) ? `${t}\n` : "";
}

var E = Symbol("clipanion/isOption");

function makeCommandOption(t) {
  return {
    ...t,
    [E]: !0
  };
}

function rerouteArguments(t, r) {
  if (void 0 === t) {
    return [ t, r ];
  }
  if ("object" == typeof t && null !== t && !Array.isArray(t)) {
    return [ void 0, t ];
  } else {
    return [ t, r ];
  }
}

function cleanValidationError(t, {mergeName: r = !1} = {}) {
  var i = t.match(/^([^:]+): (.*)$/m);
  if (!i) {
    return "validation failed";
  }
  var [, n, a] = i;
  if (r) {
    a = a[0].toLowerCase() + a.slice(1);
  }
  return a = "." !== n || !r ? `${n.replace(/^\.(\[|$)/, "$1")}: ${a}` : `: ${a}`;
}

function formatError(t, r) {
  if (1 === r.length) {
    return new UsageError(`${t}${cleanValidationError(r[0], {
      mergeName: !0
    })}`);
  } else {
    return new UsageError(`${t}:\n${r.map((t => `\n- ${cleanValidationError(t)}`)).join("")}`);
  }
}

function applyValidator(t, r, i) {
  if (void 0 === i) {
    return r;
  }
  var n = [];
  var a = [];
  var coercion = t => {
    var i = r;
    r = t;
    return coercion.bind(null, i);
  };
  if (!i(r, {
    errors: n,
    coercions: a,
    coercion
  })) {
    throw formatError(`Invalid value for ${t}`, n);
  }
  for (var [, o] of a) {
    o();
  }
  return r;
}

class Command {
  constructor() {
    this.help = !1;
  }
  static Usage(t) {
    return t;
  }
  async catch(t) {
    throw t;
  }
  async validateAndExecute() {
    var t = this.constructor.schema;
    if (Array.isArray(t)) {
      var {isDict: r, isUnknown: i, applyCascade: n} = await Promise.resolve().then((function() {
        return J;
      }));
      var a = [];
      var o = [];
      if (!n(r(i()), t)(this, {
        errors: a,
        coercions: o
      })) {
        throw formatError("Invalid option schema", a);
      }
      for (var [, s] of o) {
        s();
      }
    } else if (null != t) {
      throw new Error("Invalid command schema");
    }
    var l = await this.execute();
    if (void 0 !== l) {
      return l;
    } else {
      return 0;
    }
  }
}

Command.isOption = E;

Command.Default = [];

function debug(t) {
  if (y) {
    console.log(t);
  }
}

var C = {
  candidateUsage: null,
  requiredOptions: [],
  errorMessage: null,
  ignoreOptions: !1,
  path: [],
  positionals: [],
  options: [],
  remainder: null,
  selectedIndex: g,
  tokens: []
};

function makeStateMachine() {
  var t = {
    nodes: []
  };
  for (var r = 0; r < f.CustomNode; ++r) {
    t.nodes.push({
      dynamics: [],
      shortcuts: [],
      statics: {}
    });
  }
  return t;
}

function injectNode(t, r) {
  t.nodes.push(r);
  return t.nodes.length - 1;
}

function runMachineInternal(t, r, i = !1) {
  debug(`Running a vm on ${JSON.stringify(r)}`);
  var n = [ {
    node: f.InitialNode,
    state: {
      candidateUsage: null,
      requiredOptions: [],
      errorMessage: null,
      ignoreOptions: !1,
      options: [],
      path: [],
      positionals: [],
      remainder: null,
      selectedIndex: null,
      tokens: []
    }
  } ];
  !function debugMachine(t, {prefix: r = ""} = {}) {
    if (y) {
      debug(`${r}Nodes are:`);
      for (var i = 0; i < t.nodes.length; ++i) {
        debug(`${r}  ${i}: ${JSON.stringify(t.nodes[i])}`);
      }
    }
  }(t, {
    prefix: "  "
  });
  var a = [ d.StartOfInput, ...r ];
  for (var o = 0; o < a.length; ++o) {
    var s = a[o];
    var l = s === d.EndOfInput || s === d.EndOfPartialInput;
    var c = o - 1;
    debug(`  Processing ${JSON.stringify(s)}`);
    var u = [];
    for (var {node: g, state: p} of n) {
      debug(`    Current node is ${g}`);
      var h = t.nodes[g];
      if (g === f.ErrorNode) {
        u.push({
          node: g,
          state: p
        });
        continue;
      }
      console.assert(0 === h.shortcuts.length, "Shortcuts should have been eliminated by now");
      var v = Object.prototype.hasOwnProperty.call(h.statics, s);
      if (!i || o < a.length - 1 || v) {
        if (v) {
          var m = h.statics[s];
          for (var {to: S, reducer: w} of m) {
            u.push({
              node: S,
              state: void 0 !== w ? execute($, w, p, s, c) : p
            });
            debug(`      Static transition to ${S} found`);
          }
        } else {
          debug("      No static transition found");
        }
      } else {
        var b = !1;
        for (var x of Object.keys(h.statics)) {
          if (!x.startsWith(s)) {
            continue;
          }
          if (s === x) {
            for (var {to: E, reducer: k} of h.statics[x]) {
              u.push({
                node: E,
                state: void 0 !== k ? execute($, k, p, s, c) : p
              });
              debug(`      Static transition to ${E} found`);
            }
          } else {
            for (var {to: N} of h.statics[x]) {
              u.push({
                node: N,
                state: {
                  ...p,
                  remainder: x.slice(s.length)
                }
              });
              debug(`      Static transition to ${N} found (partial match)`);
            }
          }
          b = !0;
        }
        if (!b) {
          debug("      No partial static transition found");
        }
      }
      if (!l) {
        for (var [I, {to: T, reducer: P}] of h.dynamics) {
          if (execute(O, I, p, s, c)) {
            u.push({
              node: T,
              state: void 0 !== P ? execute($, P, p, s, c) : p
            });
            debug(`      Dynamic transition to ${T} found (via ${I})`);
          }
        }
      }
    }
    if (0 === u.length && l && 1 === r.length) {
      return [ {
        node: f.InitialNode,
        state: C
      } ];
    }
    if (0 === u.length) {
      throw new UnknownSyntaxError(r, n.filter((({node: t}) => t !== f.ErrorNode)).map((({state: t}) => ({
        usage: t.candidateUsage,
        reason: null
      }))));
    }
    if (u.every((({node: t}) => t === f.ErrorNode))) {
      throw new UnknownSyntaxError(r, u.map((({state: t}) => ({
        usage: t.candidateUsage,
        reason: t.errorMessage
      }))));
    }
    n = trimSmallerBranches(u);
  }
  if (n.length > 0) {
    debug("  Results:");
    for (var M of n) {
      debug(`    - ${M.node} -> ${JSON.stringify(M.state)}`);
    }
  } else {
    debug("  No results");
  }
  return n;
}

function runMachine(t, r, {endToken: i = d.EndOfInput} = {}) {
  return function selectBestState(t, r) {
    var i = r.filter((t => null !== t.selectedIndex));
    if (0 === i.length) {
      throw new Error;
    }
    var n = i.filter((t => t.selectedIndex === g || t.requiredOptions.every((r => r.some((r => t.options.find((t => t.name === r))))))));
    if (0 === n.length) {
      throw new UnknownSyntaxError(t, i.map((t => ({
        usage: t.candidateUsage,
        reason: null
      }))));
    }
    var a = 0;
    for (var o of n) {
      if (o.path.length > a) {
        a = o.path.length;
      }
    }
    var s = n.filter((t => t.path.length === a));
    var getPositionalCount = t => t.positionals.filter((({extra: t}) => !t)).length + t.options.length;
    var l = s.map((t => ({
      state: t,
      positionalCount: getPositionalCount(t)
    })));
    var c = 0;
    for (var {positionalCount: u} of l) {
      if (u > c) {
        c = u;
      }
    }
    var d = function aggregateHelpStates(t) {
      var r = [];
      var i = [];
      for (var n of t) {
        if (n.selectedIndex === g) {
          i.push(n);
        } else {
          r.push(n);
        }
      }
      if (i.length > 0) {
        r.push({
          ...C,
          path: findCommonPrefix(...i.map((t => t.path))),
          options: i.reduce(((t, r) => t.concat(r.options)), [])
        });
      }
      return r;
    }(l.filter((({positionalCount: t}) => t === c)).map((({state: t}) => t)));
    if (d.length > 1) {
      throw new AmbiguousSyntaxError(t, d.map((t => t.candidateUsage)));
    }
    return d[0];
  }(r, runMachineInternal(t, [ ...r, i ]).map((({state: t}) => t)));
}

function trimSmallerBranches(t) {
  var r = 0;
  for (var {state: i} of t) {
    if (i.path.length > r) {
      r = i.path.length;
    }
  }
  return t.filter((({state: t}) => t.path.length === r));
}

function findCommonPrefix(t, r, ...i) {
  if (void 0 === r) {
    return Array.from(t);
  }
  return findCommonPrefix(t.filter(((t, i) => t === r[i])), ...i);
}

function isTerminalNode(t) {
  return t === f.SuccessNode || t === f.ErrorNode;
}

function cloneTransition(t, r = 0) {
  return {
    to: !isTerminalNode(t.to) ? t.to >= f.CustomNode ? t.to + r - f.CustomNode + 1 : t.to + r : t.to,
    reducer: t.reducer
  };
}

function cloneNode(t, r = 0) {
  var i = {
    dynamics: [],
    shortcuts: [],
    statics: {}
  };
  for (var [n, a] of t.dynamics) {
    i.dynamics.push([ n, cloneTransition(a, r) ]);
  }
  for (var o of t.shortcuts) {
    i.shortcuts.push(cloneTransition(o, r));
  }
  for (var [s, l] of Object.entries(t.statics)) {
    i.statics[s] = l.map((t => cloneTransition(t, r)));
  }
  return i;
}

function registerDynamic(t, r, i, n, a) {
  t.nodes[r].dynamics.push([ i, {
    to: n,
    reducer: a
  } ]);
}

function registerShortcut(t, r, i, n) {
  t.nodes[r].shortcuts.push({
    to: i,
    reducer: n
  });
}

function registerStatic(t, r, i, n, a) {
  (!Object.prototype.hasOwnProperty.call(t.nodes[r].statics, i) ? t.nodes[r].statics[i] = [] : t.nodes[r].statics[i]).push({
    to: n,
    reducer: a
  });
}

function execute(t, r, i, n, a) {
  if (Array.isArray(r)) {
    var [o, ...s] = r;
    return t[o](i, n, a, ...s);
  } else {
    return t[r](i, n, a);
  }
}

var O = {
  always: () => !0,
  isOptionLike: (t, r) => !t.ignoreOptions && "-" !== r && r.startsWith("-"),
  isNotOptionLike: (t, r) => t.ignoreOptions || "-" === r || !r.startsWith("-"),
  isOption: (t, r, i, n) => !t.ignoreOptions && r === n,
  isBatchOption: (t, r, i, n) => !t.ignoreOptions && v.test(r) && [ ...r.slice(1) ].every((t => n.has(`-${t}`))),
  isBoundOption: (t, r, i, n, a) => {
    var o = r.match(m);
    return !t.ignoreOptions && !!o && h.test(o[1]) && n.has(o[1]) && a.filter((t => t.nameSet.includes(o[1]))).every((t => t.allowBinding));
  },
  isNegatedOption: (t, r, i, n) => !t.ignoreOptions && r === `--no-${n.slice(2)}`,
  isHelp: (t, r) => !t.ignoreOptions && p.test(r),
  isUnsupportedOption: (t, r, i, n) => !t.ignoreOptions && r.startsWith("-") && h.test(r) && !n.has(r),
  isInvalidOption: (t, r) => !t.ignoreOptions && r.startsWith("-") && !h.test(r)
};

var $ = {
  setCandidateState: (t, r, i, n) => ({
    ...t,
    ...n
  }),
  setSelectedIndex: (t, r, i, n) => ({
    ...t,
    selectedIndex: n
  }),
  pushBatch: (t, r, i, n) => {
    var a = t.options.slice();
    var o = t.tokens.slice();
    for (var s = 1; s < r.length; ++s) {
      var l = n.get(`-${r[s]}`);
      var c = 1 === s ? [ 0, 2 ] : [ s, s + 1 ];
      a.push({
        name: l,
        value: !0
      });
      o.push({
        segmentIndex: i,
        type: "option",
        option: l,
        slice: c
      });
    }
    return {
      ...t,
      options: a,
      tokens: o
    };
  },
  pushBound: (t, r, i) => {
    var [, n, a] = r.match(m);
    var o = t.options.concat({
      name: n,
      value: a
    });
    var s = t.tokens.concat([ {
      segmentIndex: i,
      type: "option",
      slice: [ 0, n.length ],
      option: n
    }, {
      segmentIndex: i,
      type: "assign",
      slice: [ n.length, n.length + 1 ]
    }, {
      segmentIndex: i,
      type: "value",
      slice: [ n.length + 1, n.length + a.length + 1 ]
    } ]);
    return {
      ...t,
      options: o,
      tokens: s
    };
  },
  pushPath: (t, r, i) => {
    var n = t.path.concat(r);
    var a = t.tokens.concat({
      segmentIndex: i,
      type: "path"
    });
    return {
      ...t,
      path: n,
      tokens: a
    };
  },
  pushPositional: (t, r, i) => {
    var n = t.positionals.concat({
      value: r,
      extra: !1
    });
    var a = t.tokens.concat({
      segmentIndex: i,
      type: "positional"
    });
    return {
      ...t,
      positionals: n,
      tokens: a
    };
  },
  pushExtra: (t, r, i) => {
    var n = t.positionals.concat({
      value: r,
      extra: !0
    });
    var a = t.tokens.concat({
      segmentIndex: i,
      type: "positional"
    });
    return {
      ...t,
      positionals: n,
      tokens: a
    };
  },
  pushExtraNoLimits: (t, r, i) => {
    var n = t.positionals.concat({
      value: r,
      extra: k
    });
    var a = t.tokens.concat({
      segmentIndex: i,
      type: "positional"
    });
    return {
      ...t,
      positionals: n,
      tokens: a
    };
  },
  pushTrue: (t, r, i, n) => {
    var a = t.options.concat({
      name: n,
      value: !0
    });
    var o = t.tokens.concat({
      segmentIndex: i,
      type: "option",
      option: n
    });
    return {
      ...t,
      options: a,
      tokens: o
    };
  },
  pushFalse: (t, r, i, n) => {
    var a = t.options.concat({
      name: n,
      value: !1
    });
    var o = t.tokens.concat({
      segmentIndex: i,
      type: "option",
      option: n
    });
    return {
      ...t,
      options: a,
      tokens: o
    };
  },
  pushUndefined: (t, r, i, n) => {
    var a = t.options.concat({
      name: r,
      value: void 0
    });
    var o = t.tokens.concat({
      segmentIndex: i,
      type: "option",
      option: r
    });
    return {
      ...t,
      options: a,
      tokens: o
    };
  },
  pushStringValue: (t, r, i) => {
    var n;
    var a = t.options[t.options.length - 1];
    var o = t.options.slice();
    var s = t.tokens.concat({
      segmentIndex: i,
      type: "value"
    });
    a.value = (null !== (n = a.value) && void 0 !== n ? n : []).concat([ r ]);
    return {
      ...t,
      options: o,
      tokens: s
    };
  },
  setStringValue: (t, r, i) => {
    var n = t.options[t.options.length - 1];
    var a = t.options.slice();
    var o = t.tokens.concat({
      segmentIndex: i,
      type: "value"
    });
    n.value = r;
    return {
      ...t,
      options: a,
      tokens: o
    };
  },
  inhibateOptions: t => ({
    ...t,
    ignoreOptions: !0
  }),
  useHelp: (t, r, i, n) => {
    var [, , a] = r.match(p);
    if (void 0 !== a) {
      return {
        ...t,
        options: [ {
          name: "-c",
          value: String(n)
        }, {
          name: "-i",
          value: a
        } ]
      };
    } else {
      return {
        ...t,
        options: [ {
          name: "-c",
          value: String(n)
        } ]
      };
    }
  },
  setError: (t, r, i, n) => {
    if (r === d.EndOfInput || r === d.EndOfPartialInput) {
      return {
        ...t,
        errorMessage: `${n}.`
      };
    } else {
      return {
        ...t,
        errorMessage: `${n} ("${r}").`
      };
    }
  },
  setOptionArityError: (t, r) => {
    var i = t.options[t.options.length - 1];
    return {
      ...t,
      errorMessage: `Not enough arguments to option ${i.name}.`
    };
  }
};

var k = Symbol();

class CommandBuilder {
  constructor(t, r) {
    this.allOptionNames = new Map;
    this.arity = {
      leading: [],
      trailing: [],
      extra: [],
      proxy: !1
    };
    this.options = [];
    this.paths = [];
    this.cliIndex = t;
    this.cliOpts = r;
  }
  addPath(t) {
    this.paths.push(t);
  }
  setArity({leading: t = this.arity.leading, trailing: r = this.arity.trailing, extra: i = this.arity.extra, proxy: n = this.arity.proxy}) {
    Object.assign(this.arity, {
      leading: t,
      trailing: r,
      extra: i,
      proxy: n
    });
  }
  addPositional({name: t = "arg", required: r = !0} = {}) {
    if (!r && this.arity.extra === k) {
      throw new Error("Optional parameters cannot be declared when using .rest() or .proxy()");
    }
    if (!r && this.arity.trailing.length > 0) {
      throw new Error("Optional parameters cannot be declared after the required trailing positional arguments");
    }
    if (!r && this.arity.extra !== k) {
      this.arity.extra.push(t);
    } else if (this.arity.extra !== k && 0 === this.arity.extra.length) {
      this.arity.leading.push(t);
    } else {
      this.arity.trailing.push(t);
    }
  }
  addRest({name: t = "arg", required: r = 0} = {}) {
    if (this.arity.extra === k) {
      throw new Error("Infinite lists cannot be declared multiple times in the same command");
    }
    if (this.arity.trailing.length > 0) {
      throw new Error("Infinite lists cannot be declared after the required trailing positional arguments");
    }
    for (var i = 0; i < r; ++i) {
      this.addPositional({
        name: t
      });
    }
    this.arity.extra = k;
  }
  addProxy({required: t = 0} = {}) {
    this.addRest({
      required: t
    });
    this.arity.proxy = !0;
  }
  addOption({names: t, description: r, arity: i = 0, hidden: n = !1, required: a = !1, allowBinding: o = !0}) {
    if (!o && i > 1) {
      throw new Error("The arity cannot be higher than 1 when the option only supports the --arg=value syntax");
    }
    if (!Number.isInteger(i)) {
      throw new Error(`The arity must be an integer, got ${i}`);
    }
    if (i < 0) {
      throw new Error(`The arity must be positive, got ${i}`);
    }
    var s = t.reduce(((t, r) => r.length > t.length ? r : t), "");
    for (var l of t) {
      this.allOptionNames.set(l, s);
    }
    this.options.push({
      preferredName: s,
      nameSet: t,
      description: r,
      arity: i,
      hidden: n,
      required: a,
      allowBinding: o
    });
  }
  setContext(t) {
    this.context = t;
  }
  usage({detailed: t = !0, inlineOptions: r = !0} = {}) {
    var i = [ this.cliOpts.binaryName ];
    var n = [];
    if (this.paths.length > 0) {
      i.push(...this.paths[0]);
    }
    if (t) {
      for (var {preferredName: a, nameSet: o, arity: s, hidden: l, description: c, required: u} of this.options) {
        if (l) {
          continue;
        }
        var d = [];
        for (var f = 0; f < s; ++f) {
          d.push(` #${f}`);
        }
        var g = `${o.join(",")}${d.join("")}`;
        if (!r && c) {
          n.push({
            preferredName: a,
            nameSet: o,
            definition: g,
            description: c,
            required: u
          });
        } else {
          i.push(u ? `<${g}>` : `[${g}]`);
        }
      }
      i.push(...this.arity.leading.map((t => `<${t}>`)));
      if (this.arity.extra === k) {
        i.push("...");
      } else {
        i.push(...this.arity.extra.map((t => `[${t}]`)));
      }
      i.push(...this.arity.trailing.map((t => `<${t}>`)));
    }
    return {
      usage: i.join(" "),
      options: n
    };
  }
  compile() {
    if (void 0 === this.context) {
      throw new Error("Assertion failed: No context attached");
    }
    var t = makeStateMachine();
    var r;
    var i = this.usage().usage;
    var n = this.options.filter((t => t.required)).map((t => t.nameSet));
    r = injectNode(t, {
      dynamics: [],
      shortcuts: [],
      statics: {}
    });
    registerStatic(t, f.InitialNode, d.StartOfInput, r, [ "setCandidateState", {
      candidateUsage: i,
      requiredOptions: n
    } ]);
    var a = this.arity.proxy ? "always" : "isNotOptionLike";
    var o = this.paths.length > 0 ? this.paths : [ [] ];
    for (var s of o) {
      var l = r;
      if (s.length > 0) {
        var c = injectNode(t, {
          dynamics: [],
          shortcuts: [],
          statics: {}
        });
        registerShortcut(t, l, c);
        this.registerOptions(t, c);
        l = c;
      }
      for (var u = 0; u < s.length; ++u) {
        var p = injectNode(t, {
          dynamics: [],
          shortcuts: [],
          statics: {}
        });
        registerStatic(t, l, s[u], p, "pushPath");
        l = p;
        if (u + 1 < s.length) {
          var h = injectNode(t, {
            dynamics: [],
            shortcuts: [],
            statics: {}
          });
          registerDynamic(t, l, "isHelp", h, [ "useHelp", this.cliIndex ]);
          registerStatic(t, h, d.EndOfInput, f.SuccessNode, [ "setSelectedIndex", g ]);
        }
      }
      if (this.arity.leading.length > 0 || !this.arity.proxy) {
        var v = injectNode(t, {
          dynamics: [],
          shortcuts: [],
          statics: {}
        });
        registerDynamic(t, l, "isHelp", v, [ "useHelp", this.cliIndex ]);
        registerDynamic(t, v, "always", v, "pushExtra");
        registerStatic(t, v, d.EndOfInput, f.SuccessNode, [ "setSelectedIndex", g ]);
        this.registerOptions(t, l);
      }
      if (this.arity.leading.length > 0) {
        registerStatic(t, l, d.EndOfInput, f.ErrorNode, [ "setError", "Not enough positional arguments" ]);
        registerStatic(t, l, d.EndOfPartialInput, f.SuccessNode, [ "setSelectedIndex", this.cliIndex ]);
      }
      var m = l;
      for (var y = 0; y < this.arity.leading.length; ++y) {
        var S = injectNode(t, {
          dynamics: [],
          shortcuts: [],
          statics: {}
        });
        if (!this.arity.proxy || y + 1 !== this.arity.leading.length) {
          this.registerOptions(t, S);
        }
        if (this.arity.trailing.length > 0 || y + 1 !== this.arity.leading.length) {
          registerStatic(t, S, d.EndOfInput, f.ErrorNode, [ "setError", "Not enough positional arguments" ]);
          registerStatic(t, S, d.EndOfPartialInput, f.SuccessNode, [ "setSelectedIndex", this.cliIndex ]);
        }
        registerDynamic(t, m, "isNotOptionLike", S, "pushPositional");
        m = S;
      }
      var w = m;
      if (this.arity.extra === k || this.arity.extra.length > 0) {
        var b = injectNode(t, {
          dynamics: [],
          shortcuts: [],
          statics: {}
        });
        registerShortcut(t, m, b);
        if (this.arity.extra === k) {
          var x = injectNode(t, {
            dynamics: [],
            shortcuts: [],
            statics: {}
          });
          if (!this.arity.proxy) {
            this.registerOptions(t, x);
          }
          registerDynamic(t, m, a, x, "pushExtraNoLimits");
          registerDynamic(t, x, a, x, "pushExtraNoLimits");
          registerShortcut(t, x, b);
        } else {
          for (var E = 0; E < this.arity.extra.length; ++E) {
            var C = injectNode(t, {
              dynamics: [],
              shortcuts: [],
              statics: {}
            });
            if (!this.arity.proxy || E > 0) {
              this.registerOptions(t, C);
            }
            registerDynamic(t, w, a, C, "pushExtra");
            registerShortcut(t, C, b);
            w = C;
          }
        }
        w = b;
      }
      if (this.arity.trailing.length > 0) {
        registerStatic(t, w, d.EndOfInput, f.ErrorNode, [ "setError", "Not enough positional arguments" ]);
        registerStatic(t, w, d.EndOfPartialInput, f.SuccessNode, [ "setSelectedIndex", this.cliIndex ]);
      }
      var O = w;
      for (var $ = 0; $ < this.arity.trailing.length; ++$) {
        var N = injectNode(t, {
          dynamics: [],
          shortcuts: [],
          statics: {}
        });
        if (!this.arity.proxy) {
          this.registerOptions(t, N);
        }
        if ($ + 1 < this.arity.trailing.length) {
          registerStatic(t, N, d.EndOfInput, f.ErrorNode, [ "setError", "Not enough positional arguments" ]);
          registerStatic(t, N, d.EndOfPartialInput, f.SuccessNode, [ "setSelectedIndex", this.cliIndex ]);
        }
        registerDynamic(t, O, "isNotOptionLike", N, "pushPositional");
        O = N;
      }
      registerDynamic(t, O, a, f.ErrorNode, [ "setError", "Extraneous positional argument" ]);
      registerStatic(t, O, d.EndOfInput, f.SuccessNode, [ "setSelectedIndex", this.cliIndex ]);
      registerStatic(t, O, d.EndOfPartialInput, f.SuccessNode, [ "setSelectedIndex", this.cliIndex ]);
    }
    return {
      machine: t,
      context: this.context
    };
  }
  registerOptions(t, r) {
    registerDynamic(t, r, [ "isOption", "--" ], r, "inhibateOptions");
    registerDynamic(t, r, [ "isBatchOption", this.allOptionNames ], r, [ "pushBatch", this.allOptionNames ]);
    registerDynamic(t, r, [ "isBoundOption", this.allOptionNames, this.options ], r, "pushBound");
    registerDynamic(t, r, [ "isUnsupportedOption", this.allOptionNames ], f.ErrorNode, [ "setError", "Unsupported option name" ]);
    registerDynamic(t, r, [ "isInvalidOption" ], f.ErrorNode, [ "setError", "Invalid option name" ]);
    for (var i of this.options) {
      if (0 === i.arity) {
        for (var n of i.nameSet) {
          registerDynamic(t, r, [ "isOption", n ], r, [ "pushTrue", i.preferredName ]);
          if (n.startsWith("--") && !n.startsWith("--no-")) {
            registerDynamic(t, r, [ "isNegatedOption", n ], r, [ "pushFalse", i.preferredName ]);
          }
        }
      } else {
        var a = injectNode(t, {
          dynamics: [],
          shortcuts: [],
          statics: {}
        });
        for (var o of i.nameSet) {
          registerDynamic(t, r, [ "isOption", o ], a, [ "pushUndefined", i.preferredName ]);
        }
        for (var s = 0; s < i.arity; ++s) {
          var l = injectNode(t, {
            dynamics: [],
            shortcuts: [],
            statics: {}
          });
          registerStatic(t, a, d.EndOfInput, f.ErrorNode, "setOptionArityError");
          registerStatic(t, a, d.EndOfPartialInput, f.ErrorNode, "setOptionArityError");
          registerDynamic(t, a, "isOptionLike", f.ErrorNode, "setOptionArityError");
          registerDynamic(t, a, "isNotOptionLike", l, 1 === i.arity ? "setStringValue" : "pushStringValue");
          a = l;
        }
        registerShortcut(t, a, r);
      }
    }
  }
}

class CliBuilder {
  constructor({binaryName: t = "..."} = {}) {
    this.builders = [];
    this.opts = {
      binaryName: t
    };
  }
  static build(t, r = {}) {
    return new CliBuilder(r).commands(t).compile();
  }
  getBuilderByIndex(t) {
    if (!(t >= 0 && t < this.builders.length)) {
      throw new Error(`Assertion failed: Out-of-bound command index (${t})`);
    }
    return this.builders[t];
  }
  commands(t) {
    for (var r of t) {
      r(this.command());
    }
    return this;
  }
  command() {
    var t = new CommandBuilder(this.builders.length, this.opts);
    this.builders.push(t);
    return t;
  }
  compile() {
    var t = [];
    var r = [];
    for (var i of this.builders) {
      var {machine: n, context: a} = i.compile();
      t.push(n);
      r.push(a);
    }
    var o = function makeAnyOfMachine(t) {
      var r = makeStateMachine();
      var i = [];
      var n = r.nodes.length;
      for (var a of t) {
        i.push(n);
        for (var o = 0; o < a.nodes.length; ++o) {
          if (!isTerminalNode(o)) {
            r.nodes.push(cloneNode(a.nodes[o], n));
          }
        }
        n += a.nodes.length - f.CustomNode + 1;
      }
      for (var s of i) {
        registerShortcut(r, f.InitialNode, s);
      }
      return r;
    }(t);
    !function simplifyMachine(t) {
      var r = new Set;
      var process = i => {
        if (r.has(i)) {
          return;
        }
        r.add(i);
        var n = t.nodes[i];
        for (var a of Object.values(n.statics)) {
          for (var {to: o} of a) {
            process(o);
          }
        }
        for (var [, {to: s}] of n.dynamics) {
          process(s);
        }
        for (var {to: l} of n.shortcuts) {
          process(l);
        }
        var c = new Set(n.shortcuts.map((({to: t}) => t)));
        while (n.shortcuts.length > 0) {
          var {to: u} = n.shortcuts.shift();
          var d = t.nodes[u];
          for (var [f, g] of Object.entries(d.statics)) {
            var p = !Object.prototype.hasOwnProperty.call(n.statics, f) ? n.statics[f] = [] : n.statics[f];
            var _loop = function(t) {
              if (!p.some((({to: r}) => t.to === r))) {
                p.push(t);
              }
            };
            for (var h of g) {
              _loop(h);
            }
          }
          var _loop2 = function(t, r) {
            if (!n.dynamics.some((([i, {to: n}]) => t === i && r.to === n))) {
              n.dynamics.push([ t, r ]);
            }
          };
          for (var [v, m] of d.dynamics) {
            _loop2(v, m);
          }
          for (var y of d.shortcuts) {
            if (!c.has(y.to)) {
              n.shortcuts.push(y);
              c.add(y.to);
            }
          }
        }
      };
      process(f.InitialNode);
    }(o);
    return {
      machine: o,
      contexts: r,
      process: (t, {partial: r} = {}) => runMachine(o, t, {
        endToken: r ? d.EndOfPartialInput : d.EndOfInput
      })
    };
  }
}

class HelpCommand extends Command {
  constructor(t) {
    super();
    this.contexts = t;
    this.commands = [];
  }
  static from(t, r) {
    var i = new HelpCommand(r);
    i.path = t.path;
    for (var n of t.options) {
      switch (n.name) {
       case "-c":
        i.commands.push(Number(n.value));
        break;

       case "-i":
        i.index = Number(n.value);
      }
    }
    return i;
  }
  async execute() {
    var t = this.commands;
    if (void 0 !== this.index && this.index >= 0 && this.index < t.length) {
      t = [ t[this.index] ];
    }
    if (0 === t.length) {
      this.context.stdout.write(this.cli.usage());
    } else if (1 === t.length) {
      this.context.stdout.write(this.cli.usage(this.contexts[t[0]].commandClass, {
        detailed: !0
      }));
    } else if (t.length > 1) {
      this.context.stdout.write("Multiple commands match your selection:\n");
      this.context.stdout.write("\n");
      var r = 0;
      for (var i of this.commands) {
        this.context.stdout.write(this.cli.usage(this.contexts[i].commandClass, {
          prefix: (r++ + ". ").padStart(5)
        }));
      }
      this.context.stdout.write("\n");
      this.context.stdout.write("Run again with -h=<index> to see the longer details of any of those commands.\n");
    }
  }
}

var N = Symbol("clipanion/errorCommand");

class Cli {
  constructor({binaryLabel: t, binaryName: r = "...", binaryVersion: i, enableCapture: n = !1, enableColors: a} = {}) {
    this.registrations = new Map;
    this.builder = new CliBuilder({
      binaryName: r
    });
    this.binaryLabel = t;
    this.binaryName = r;
    this.binaryVersion = i;
    this.enableCapture = n;
    this.enableColors = a;
  }
  static from(t, r = {}) {
    var i = new Cli(r);
    var n = Array.isArray(t) ? t : [ t ];
    for (var a of n) {
      i.register(a);
    }
    return i;
  }
  register(t) {
    var r;
    var i = new Map;
    var n = new t;
    for (var a in n) {
      var o = n[a];
      if ("object" == typeof o && null !== o && o[Command.isOption]) {
        i.set(a, o);
      }
    }
    var s = this.builder.command();
    var l = s.cliIndex;
    var c = null !== (r = t.paths) && void 0 !== r ? r : n.paths;
    if (void 0 !== c) {
      for (var u of c) {
        s.addPath(u);
      }
    }
    this.registrations.set(t, {
      specs: i,
      builder: s,
      index: l
    });
    for (var [d, {definition: f}] of i.entries()) {
      f(s, d);
    }
    s.setContext({
      commandClass: t
    });
  }
  process(t, r) {
    var {input: i, context: n, partial: a} = "object" == typeof t && Array.isArray(t) ? {
      input: t,
      context: r
    } : t;
    var {contexts: o, process: s} = this.builder.compile();
    var l = s(i, {
      partial: a
    });
    var c = {
      ...Cli.defaultContext,
      ...n
    };
    if (l.selectedIndex === g) {
      var u = HelpCommand.from(l, o);
      u.context = c;
      u.tokens = l.tokens;
      return u;
    } else {
      var {commandClass: d} = o[l.selectedIndex];
      var f = this.registrations.get(d);
      if (void 0 === f) {
        throw new Error("Assertion failed: Expected the command class to have been registered.");
      }
      var p = new d;
      p.context = c;
      p.tokens = l.tokens;
      p.path = l.path;
      try {
        for (var [h, {transformer: v}] of f.specs.entries()) {
          p[h] = v(f.builder, h, l, c);
        }
        return p;
      } catch (t) {
        t[N] = p;
        throw t;
      }
    }
  }
  async run(t, r) {
    var i, n;
    var a;
    var o = {
      ...Cli.defaultContext,
      ...r
    };
    var s = null !== (i = this.enableColors) && void 0 !== i ? i : o.colorDepth > 1;
    if (!Array.isArray(t)) {
      a = t;
    } else {
      try {
        a = this.process(t, o);
      } catch (t) {
        o.stdout.write(this.error(t, {
          colored: s
        }));
        return 1;
      }
    }
    if (a.help) {
      o.stdout.write(this.usage(a, {
        colored: s,
        detailed: !0
      }));
      return 0;
    }
    a.context = o;
    a.cli = {
      binaryLabel: this.binaryLabel,
      binaryName: this.binaryName,
      binaryVersion: this.binaryVersion,
      enableCapture: this.enableCapture,
      enableColors: this.enableColors,
      definitions: () => this.definitions(),
      definition: t => this.definition(t),
      error: (t, r) => this.error(t, r),
      format: t => this.format(t),
      process: (t, r) => this.process(t, {
        ...o,
        ...r
      }),
      run: (t, r) => this.run(t, {
        ...o,
        ...r
      }),
      usage: (t, r) => this.usage(t, r)
    };
    var l = this.enableCapture ? null !== (n = function getCaptureActivator() {
      throw new Error("The enableCapture option cannot be used from within a browser environment");
    }()) && void 0 !== n ? n : noopCaptureActivator : noopCaptureActivator;
    var c;
    try {
      c = await l((() => a.validateAndExecute().catch((t => a.catch(t).then((() => 0))))));
    } catch (t) {
      o.stdout.write(this.error(t, {
        colored: s,
        command: a
      }));
      return 1;
    }
    return c;
  }
  async runExit(t, r) {
    process.exitCode = await this.run(t, r);
  }
  definition(t, {colored: r = !1} = {}) {
    if (!t.usage) {
      return null;
    }
    var {usage: i} = this.getUsageByRegistration(t, {
      detailed: !1
    });
    var {usage: n, options: a} = this.getUsageByRegistration(t, {
      detailed: !0,
      inlineOptions: !1
    });
    return {
      path: i,
      usage: n,
      category: void 0 !== t.usage.category ? formatMarkdownish(t.usage.category, {
        format: this.format(r),
        paragraphs: !1
      }) : void 0,
      description: void 0 !== t.usage.description ? formatMarkdownish(t.usage.description, {
        format: this.format(r),
        paragraphs: !1
      }) : void 0,
      details: void 0 !== t.usage.details ? formatMarkdownish(t.usage.details, {
        format: this.format(r),
        paragraphs: !0
      }) : void 0,
      examples: void 0 !== t.usage.examples ? t.usage.examples.map((([t, i]) => [ formatMarkdownish(t, {
        format: this.format(r),
        paragraphs: !1
      }), i.replace(/\$0/g, this.binaryName) ])) : void 0,
      options: a
    };
  }
  definitions({colored: t = !1} = {}) {
    var r = [];
    for (var i of this.registrations.keys()) {
      var n = this.definition(i, {
        colored: t
      });
      if (!n) {
        continue;
      }
      r.push(n);
    }
    return r;
  }
  usage(t = null, {colored: r, detailed: i = !1, prefix: n = "$ "} = {}) {
    var a;
    if (null === t) {
      for (var o of this.registrations.keys()) {
        var s = o.paths;
        var l = void 0 !== o.usage;
        var c = !s || 0 === s.length || 1 === s.length && 0 === s[0].length || (null !== (a = null == s ? void 0 : s.some((t => 0 === t.length))) && void 0 !== a ? a : !1);
        if (c) {
          if (t) {
            t = null;
            break;
          } else {
            t = o;
          }
        } else if (l) {
          t = null;
          continue;
        }
      }
      if (t) {
        i = !0;
      }
    }
    var u = null !== t && t instanceof Command ? t.constructor : t;
    var d = "";
    if (!u) {
      var f = new Map;
      for (var [g, {index: p}] of this.registrations.entries()) {
        if (void 0 === g.usage) {
          continue;
        }
        var h = void 0 !== g.usage.category ? formatMarkdownish(g.usage.category, {
          format: this.format(r),
          paragraphs: !1
        }) : null;
        var v = f.get(h);
        if (void 0 === v) {
          f.set(h, v = []);
        }
        var {usage: m} = this.getUsageByIndex(p);
        v.push({
          commandClass: g,
          usage: m
        });
      }
      var y = Array.from(f.keys()).sort(((t, r) => {
        if (null === t) {
          return -1;
        }
        if (null === r) {
          return 1;
        }
        return t.localeCompare(r, "en", {
          usage: "sort",
          caseFirst: "upper"
        });
      }));
      var S = void 0 !== this.binaryLabel;
      var w = void 0 !== this.binaryVersion;
      if (S || w) {
        if (S && w) {
          d += `${this.format(r).header(`${this.binaryLabel} - ${this.binaryVersion}`)}\n\n`;
        } else if (S) {
          d += `${this.format(r).header(`${this.binaryLabel}`)}\n`;
        } else {
          d += `${this.format(r).header(`${this.binaryVersion}`)}\n`;
        }
        d += `  ${this.format(r).bold(n)}${this.binaryName} <command>\n`;
      } else {
        d += `${this.format(r).bold(n)}${this.binaryName} <command>\n`;
      }
      for (var b of y) {
        var x = f.get(b).slice().sort(((t, r) => t.usage.localeCompare(r.usage, "en", {
          usage: "sort",
          caseFirst: "upper"
        })));
        var E = null !== b ? b.trim() : "General commands";
        d += "\n";
        d += `${this.format(r).header(`${E}`)}\n`;
        for (var {commandClass: C, usage: O} of x) {
          var $ = C.usage.description || "undocumented";
          d += "\n";
          d += `  ${this.format(r).bold(O)}\n`;
          d += `    ${formatMarkdownish($, {
            format: this.format(r),
            paragraphs: !1
          })}`;
        }
      }
      d += "\n";
      d += formatMarkdownish("You can also print more details about any of these commands by calling them with the `-h,--help` flag right after the command name.", {
        format: this.format(r),
        paragraphs: !0
      });
    } else if (!i) {
      var {usage: k} = this.getUsageByRegistration(u);
      d += `${this.format(r).bold(n)}${k}\n`;
    } else {
      var {description: N = "", details: I = "", examples: T = []} = u.usage || {};
      if ("" !== N) {
        d += formatMarkdownish(N, {
          format: this.format(r),
          paragraphs: !1
        }).replace(/^./, (t => t.toUpperCase()));
        d += "\n";
      }
      if ("" !== I || T.length > 0) {
        d += `${this.format(r).header("Usage")}\n`;
        d += "\n";
      }
      var {usage: P, options: M} = this.getUsageByRegistration(u, {
        inlineOptions: !1
      });
      d += `${this.format(r).bold(n)}${P}\n`;
      if (M.length > 0) {
        d += "\n";
        d += `${this.format(r).header("Options")}\n`;
        var B = M.reduce(((t, r) => Math.max(t, r.definition.length)), 0);
        d += "\n";
        for (var {definition: F, description: A} of M) {
          d += `  ${this.format(r).bold(F.padEnd(B))}    ${formatMarkdownish(A, {
            format: this.format(r),
            paragraphs: !1
          })}`;
        }
      }
      if ("" !== I) {
        d += "\n";
        d += `${this.format(r).header("Details")}\n`;
        d += "\n";
        d += formatMarkdownish(I, {
          format: this.format(r),
          paragraphs: !0
        });
      }
      if (T.length > 0) {
        d += "\n";
        d += `${this.format(r).header("Examples")}\n`;
        for (var [L, R] of T) {
          d += "\n";
          d += formatMarkdownish(L, {
            format: this.format(r),
            paragraphs: !1
          });
          d += `${R.replace(/^/m, `  ${this.format(r).bold(n)}`).replace(/\$0/g, this.binaryName)}\n`;
        }
      }
    }
    return d;
  }
  error(t, r) {
    var i;
    var {colored: n, command: a = (null !== (i = t[N]) && void 0 !== i ? i : null)} = void 0 === r ? {} : r;
    if (!t || "object" != typeof t || !("stack" in t)) {
      t = new Error(`Execution failed with a non-error rejection (rejected value: ${JSON.stringify(t)})`);
    }
    var o = "";
    var s = t.name.replace(/([a-z])([A-Z])/g, "$1 $2");
    if ("Error" === s) {
      s = "Internal Error";
    }
    o += `${this.format(n).error(s)}: ${t.message}\n`;
    var l = t.clipanion;
    if (void 0 !== l) {
      if ("usage" === l.type) {
        o += "\n";
        o += this.usage(a);
      }
    } else if (t.stack) {
      o += `${t.stack.replace(/^.*\n/, "")}\n`;
    }
    return o;
  }
  format(t) {
    var r;
    return (null !== (r = null != t ? t : this.enableColors) && void 0 !== r ? r : Cli.defaultContext.colorDepth > 1) ? b : x;
  }
  getUsageByRegistration(t, r) {
    var i = this.registrations.get(t);
    if (void 0 === i) {
      throw new Error("Assertion failed: Unregistered command");
    }
    return this.getUsageByIndex(i.index, r);
  }
  getUsageByIndex(t, r) {
    return this.builder.getBuilderByIndex(t).usage(r);
  }
}

Cli.defaultContext = {
  env: process.env,
  stdin: process.stdin,
  stdout: process.stdout,
  stderr: process.stderr,
  colorDepth: 1
};

function noopCaptureActivator(t) {
  return t();
}

function Boolean(t, r, i) {
  var [n, a] = rerouteArguments(r, null != i ? i : {});
  var o = t.split(",");
  var s = new Set(o);
  return makeCommandOption({
    definition(t) {
      t.addOption({
        names: o,
        allowBinding: !1,
        arity: 0,
        hidden: a.hidden,
        description: a.description,
        required: a.required
      });
    },
    transformer(t, r, i) {
      var a = n;
      for (var {name: o, value: l} of i.options) {
        if (!s.has(o)) {
          continue;
        }
        a = l;
      }
      return a;
    }
  });
}

function String$1(t, ...r) {
  if ("string" == typeof t) {
    return function StringOption(t, r, i) {
      var [n, a] = rerouteArguments(r, null != i ? i : {});
      var {arity: o = 1} = a;
      var s = t.split(",");
      var l = new Set(s);
      return makeCommandOption({
        definition(t) {
          t.addOption({
            names: s,
            arity: a.tolerateBoolean ? 0 : o,
            hidden: a.hidden,
            description: a.description,
            required: a.required
          });
        },
        transformer(t, r, i, o) {
          var s;
          var c = n;
          if (void 0 !== a.env && o.env[a.env]) {
            s = a.env;
            c = o.env[a.env];
          }
          for (var {name: u, value: d} of i.options) {
            if (!l.has(u)) {
              continue;
            }
            s = u;
            c = d;
          }
          if ("string" == typeof c) {
            return applyValidator(null != s ? s : r, c, a.validator);
          } else {
            return c;
          }
        }
      });
    }(t, ...r);
  } else {
    return function StringPositional(t = {}) {
      var {required: r = !0} = t;
      return makeCommandOption({
        definition(r, i) {
          var n;
          r.addPositional({
            name: null !== (n = t.name) && void 0 !== n ? n : i,
            required: t.required
          });
        },
        transformer(i, n, a) {
          var o;
          for (var s = 0; s < a.positionals.length; ++s) {
            if (a.positionals[s].extra === k) {
              continue;
            }
            if (r && !0 === a.positionals[s].extra) {
              continue;
            }
            if (!r && !1 === a.positionals[s].extra) {
              continue;
            }
            var [l] = a.positionals.splice(s, 1);
            return applyValidator(null !== (o = t.name) && void 0 !== o ? o : n, l.value, t.validator);
          }
          return;
        }
      });
    }(t);
  }
}

var I = function(t) {
  t[t.VUE_SUPPORT = 0] = "VUE_SUPPORT";
  t[t.SVELTE_SUPPORT = 1] = "SVELTE_SUPPORT";
  t[t.UNKNOWN_EXTERNAL_FILE = 2] = "UNKNOWN_EXTERNAL_FILE";
  return t;
}({});

class TadaError extends Error {
  static isTadaError(t) {
    return !("object" != typeof t || !t || !("name" in t) || "TadaError" !== t.name);
  }
  constructor(t, r) {
    super(r);
    this.code = t;
    this.name = "TadaError";
  }
}

var exitCode = () => {
  var {exitCode: t} = process;
  return "string" == typeof t ? parseInt(t, 10) || 0 : t;
};

var e = () => {};

function start(t) {
  return {
    tag: 0,
    0: t
  };
}

function push(t) {
  return {
    tag: 1,
    0: t
  };
}

var asyncIteratorSymbol = () => "function" == typeof Symbol && Symbol.asyncIterator || "@@asyncIterator";

var identity = t => t;

function concatMap(t) {
  return r => i => {
    var n = [];
    var a = e;
    var o = e;
    var s = !1;
    var l = !1;
    var c = !1;
    var u = !1;
    function applyInnerSource(r) {
      c = !0;
      r((r => {
        if (0 === r) {
          if (c) {
            c = !1;
            if (n.length) {
              applyInnerSource(t(n.shift()));
            } else if (u) {
              i(0);
            } else if (!s) {
              s = !0;
              a(0);
            }
          }
        } else if (0 === r.tag) {
          l = !1;
          (o = r[0])(0);
        } else if (c) {
          i(r);
          if (l) {
            l = !1;
          } else {
            o(0);
          }
        }
      }));
    }
    r((r => {
      if (u) {} else if (0 === r) {
        u = !0;
        if (!c && !n.length) {
          i(0);
        }
      } else if (0 === r.tag) {
        a = r[0];
      } else {
        s = !1;
        if (c) {
          n.push(r[0]);
        } else {
          applyInnerSource(t(r[0]));
        }
      }
    }));
    i(start((t => {
      if (1 === t) {
        if (!u) {
          u = !0;
          a(1);
        }
        if (c) {
          c = !1;
          o(1);
        }
      } else {
        if (!u && !s) {
          s = !0;
          a(0);
        }
        if (c && !l) {
          l = !0;
          o(0);
        }
      }
    })));
  };
}

function concat(t) {
  return function concatAll(t) {
    return concatMap(identity)(t);
  }(T(t));
}

function filter(t) {
  return r => i => {
    var n = e;
    r((r => {
      if (0 === r) {
        i(0);
      } else if (0 === r.tag) {
        n = r[0];
        i(r);
      } else if (!t(r[0])) {
        n(0);
      } else {
        i(r);
      }
    }));
  };
}

function map(t) {
  return r => i => r((r => {
    if (0 === r || 0 === r.tag) {
      i(r);
    } else {
      i(push(t(r[0])));
    }
  }));
}

function mergeMap(t) {
  return r => i => {
    var n = [];
    var a = e;
    var o = !1;
    var s = !1;
    r((r => {
      if (s) {} else if (0 === r) {
        s = !0;
        if (!n.length) {
          i(0);
        }
      } else if (0 === r.tag) {
        a = r[0];
      } else {
        o = !1;
        !function applyInnerSource(t) {
          var r = e;
          t((t => {
            if (0 === t) {
              if (n.length) {
                var l = n.indexOf(r);
                if (l > -1) {
                  (n = n.slice()).splice(l, 1);
                }
                if (!n.length) {
                  if (s) {
                    i(0);
                  } else if (!o) {
                    o = !0;
                    a(0);
                  }
                }
              }
            } else if (0 === t.tag) {
              n.push(r = t[0]);
              r(0);
            } else if (n.length) {
              i(t);
              r(0);
            }
          }));
        }(t(r[0]));
        if (!o) {
          o = !0;
          a(0);
        }
      }
    }));
    i(start((t => {
      if (1 === t) {
        if (!s) {
          s = !0;
          a(1);
        }
        for (var r = 0, i = n, l = n.length; r < l; r++) {
          i[r](1);
        }
        n.length = 0;
      } else {
        if (!s && !o) {
          o = !0;
          a(0);
        } else {
          o = !1;
        }
        for (var c = 0, u = n, d = n.length; c < d; c++) {
          u[c](0);
        }
      }
    })));
  };
}

function merge(t) {
  return function mergeAll(t) {
    return mergeMap(identity)(t);
  }(T(t));
}

function onEnd(t) {
  return r => i => {
    var n = !1;
    r((r => {
      if (n) {} else if (0 === r) {
        n = !0;
        i(0);
        t();
      } else if (0 === r.tag) {
        var a = r[0];
        i(start((r => {
          if (1 === r) {
            n = !0;
            a(1);
            t();
          } else {
            a(r);
          }
        })));
      } else {
        i(r);
      }
    }));
  };
}

function onPush(t) {
  return r => i => {
    var n = !1;
    r((r => {
      if (n) {} else if (0 === r) {
        n = !0;
        i(0);
      } else if (0 === r.tag) {
        var a = r[0];
        i(start((t => {
          if (1 === t) {
            n = !0;
          }
          a(t);
        })));
      } else {
        t(r[0]);
        i(r);
      }
    }));
  };
}

function share(t) {
  var r = [];
  var i = e;
  var n = !1;
  return a => {
    r.push(a);
    if (1 === r.length) {
      t((t => {
        if (0 === t) {
          for (var a = 0, o = r, s = r.length; a < s; a++) {
            o[a](0);
          }
          r.length = 0;
        } else if (0 === t.tag) {
          i = t[0];
        } else {
          n = !1;
          for (var l = 0, c = r, u = r.length; l < u; l++) {
            c[l](t);
          }
        }
      }));
    }
    a(start((t => {
      if (1 === t) {
        var o = r.indexOf(a);
        if (o > -1) {
          (r = r.slice()).splice(o, 1);
        }
        if (!r.length) {
          i(1);
        }
      } else if (!n) {
        n = !0;
        i(0);
      }
    })));
  };
}

function takeUntil(t) {
  return r => i => {
    var n = e;
    var a = e;
    var o = !1;
    r((r => {
      if (o) {} else if (0 === r) {
        o = !0;
        a(1);
        i(0);
      } else if (0 === r.tag) {
        n = r[0];
        t((t => {
          if (0 === t) {} else if (0 === t.tag) {
            (a = t[0])(0);
          } else {
            o = !0;
            a(1);
            n(1);
            i(0);
          }
        }));
      } else {
        i(r);
      }
    }));
    i(start((t => {
      if (1 === t && !o) {
        o = !0;
        n(1);
        a(1);
      } else if (!o) {
        n(0);
      }
    })));
  };
}

function fromAsyncIterable(t) {
  return r => {
    var i = t[asyncIteratorSymbol()] && t[asyncIteratorSymbol()]() || t;
    var n = !1;
    var a = !1;
    var o = !1;
    var s;
    r(start((async t => {
      if (1 === t) {
        n = !0;
        if (i.return) {
          i.return();
        }
      } else if (a) {
        o = !0;
      } else {
        for (o = a = !0; o && !n; ) {
          if ((s = await i.next()).done) {
            n = !0;
            if (i.return) {
              await i.return();
            }
            r(0);
          } else {
            try {
              o = !1;
              r(push(s.value));
            } catch (t) {
              if (i.throw) {
                if (n = !!(await i.throw(t)).done) {
                  r(0);
                }
              } else {
                throw t;
              }
            }
          }
        }
        a = !1;
      }
    })));
  };
}

var T = function fromIterable(t) {
  if (t[Symbol.asyncIterator]) {
    return fromAsyncIterable(t);
  }
  return r => {
    var i = t[Symbol.iterator]();
    var n = !1;
    var a = !1;
    var o = !1;
    var s;
    r(start((t => {
      if (1 === t) {
        n = !0;
        if (i.return) {
          i.return();
        }
      } else if (a) {
        o = !0;
      } else {
        for (o = a = !0; o && !n; ) {
          if ((s = i.next()).done) {
            n = !0;
            if (i.return) {
              i.return();
            }
            r(0);
          } else {
            try {
              o = !1;
              r(push(s.value));
            } catch (t) {
              if (i.throw) {
                if (n = !!i.throw(t).done) {
                  r(0);
                }
              } else {
                throw t;
              }
            }
          }
        }
        a = !1;
      }
    })));
  };
};

function fromValue(t) {
  return r => {
    var i = !1;
    r(start((n => {
      if (1 === n) {
        i = !0;
      } else if (!i) {
        i = !0;
        r(push(t));
        r(0);
      }
    })));
  };
}

function make(t) {
  return r => {
    var i = !1;
    var n = t({
      next(t) {
        if (!i) {
          r(push(t));
        }
      },
      complete() {
        if (!i) {
          i = !0;
          r(0);
        }
      }
    });
    r(start((t => {
      if (1 === t && !i) {
        i = !0;
        n();
      }
    })));
  };
}

var never = t => {
  t(start(e));
};

function interval(t) {
  return make((r => {
    var i = 0;
    var n = setInterval((() => r.next(i++)), t);
    return () => clearInterval(n);
  }));
}

function toPromise(t) {
  return new Promise((r => {
    var i = e;
    var n;
    t((t => {
      if (0 === t) {
        Promise.resolve(n).then(r);
      } else if (0 === t.tag) {
        (i = t[0])(0);
      } else {
        n = t[0];
        i(0);
      }
    }));
  }));
}

var pipe = (...t) => {
  var r = t[0];
  for (var i = 1, n = t.length; i < n; i++) {
    r = t[i](r);
  }
  return r;
};

var P = !1;

function _setColor(t) {
  P = t;
}

var M = function(t) {
  t.InsertChars = "@";
  t.ScrollLeft = "SP@";
  t.Up = "A";
  t.ScrollRight = "SPA";
  t.Down = "B";
  t.Forward = "C";
  t.Backward = "D";
  t.NextLine = "E";
  t.PrevLine = "F";
  t.ToColumn = "G";
  t.ToPosition = "H";
  t.Tab = "I";
  t.Erase = "J";
  t.EraseLine = "K";
  t.InsertLines = "L";
  t.DeleteLines = "M";
  t.DeleteChars = "P";
  t.ScrollUp = "S";
  t.ScrollDown = "T";
  t.EraseChars = "X";
  t.TabBackwards = "Z";
  t.RepeatChar = "b";
  t.ToRow = "d";
  t.DownLine = "e";
  t.TabClear = "g";
  t.SetMode = "h";
  t.UnsetMode = "l";
  t.SetPrivateMode = "?h";
  t.UnsetPrivateMode = "?l";
  t.ResetPrivateMode = "?r";
  t.Style = "m";
  t.Reset = "!p";
  t.Protect = '"q';
  t.Cursor = "SPq";
  t.SetMargin = "r";
  t.SaveCursor = "s";
  t.RestoreCursor = "u";
  t.InsertColumns = "'}";
  t.DeleteColumns = "'~";
  return t;
}(M || {});

var B = function(t) {
  t[t.Forward = 0] = "Forward";
  t[t.Backward = 1] = "Backward";
  t[t.All = 2] = "All";
  return t;
}(B || {});

var F = function(t) {
  t[t.Insert = 4] = "Insert";
  t[t.AutomaticNewline = 20] = "AutomaticNewline";
  return t;
}(F || {});

var A = function(t) {
  t[t.AppCursorKeys = 1] = "AppCursorKeys";
  t[t.USASCII = 2] = "USASCII";
  t[t.Column132 = 3] = "Column132";
  t[t.OriginMode = 6] = "OriginMode";
  t[t.AutoWrap = 7] = "AutoWrap";
  t[t.AutoRepeat = 8] = "AutoRepeat";
  t[t.X10Mouse = 9] = "X10Mouse";
  t[t.BlinkCursor = 12] = "BlinkCursor";
  t[t.ShowCursor = 25] = "ShowCursor";
  t[t.ReverseWrapAround = 45] = "ReverseWrapAround";
  t[t.AlternativeScreenBuffer = 47] = "AlternativeScreenBuffer";
  t[t.AppKeypad = 66] = "AppKeypad";
  t[t.X11Mouse = 1e3] = "X11Mouse";
  t[t.CellMotionMouseTracking = 1002] = "CellMotionMouseTracking";
  t[t.AllMotionMouseTracking = 1003] = "AllMotionMouseTracking";
  t[t.FocusEvents = 1004] = "FocusEvents";
  t[t.Utf8Mouse = 1005] = "Utf8Mouse";
  t[t.SGRMouse = 1006] = "SGRMouse";
  t[t.UrxvtMouse = 1015] = "UrxvtMouse";
  t[t.SGRPixelsMouse = 1016] = "SGRPixelsMouse";
  t[t.SaveCursor = 1048] = "SaveCursor";
  t[t.BracketedPaste = 2004] = "BracketedPaste";
  return t;
}(A || {});

var L = function(t) {
  t[t.Reset = 0] = "Reset";
  t[t.Bold = 1] = "Bold";
  t[t.Faint = 2] = "Faint";
  t[t.Italic = 3] = "Italic";
  t[t.Underline = 4] = "Underline";
  t[t.Blink = 5] = "Blink";
  t[t.RapidBlink = 6] = "RapidBlink";
  t[t.Invert = 7] = "Invert";
  t[t.Invisible = 8] = "Invisible";
  t[t.Strikethrough = 9] = "Strikethrough";
  t[t.DoubleUnderlined = 21] = "DoubleUnderlined";
  t[t.Normal = 22] = "Normal";
  t[t.NoItalic = 23] = "NoItalic";
  t[t.NoUnderline = 24] = "NoUnderline";
  t[t.NoBlink = 25] = "NoBlink";
  t[t.NoInvert = 27] = "NoInvert";
  t[t.Visible = 28] = "Visible";
  t[t.NoStrikethrough = 29] = "NoStrikethrough";
  t[t.Black = 30] = "Black";
  t[t.Red = 31] = "Red";
  t[t.Green = 32] = "Green";
  t[t.Yellow = 33] = "Yellow";
  t[t.Blue = 34] = "Blue";
  t[t.Magenta = 35] = "Magenta";
  t[t.Cyan = 36] = "Cyan";
  t[t.White = 37] = "White";
  t[t.Foreground = 39] = "Foreground";
  t[t.OnBlack = 40] = "OnBlack";
  t[t.OnRed = 41] = "OnRed";
  t[t.OnGreen = 42] = "OnGreen";
  t[t.OnYellow = 43] = "OnYellow";
  t[t.OnBlue = 44] = "OnBlue";
  t[t.OnMagenta = 45] = "OnMagenta";
  t[t.OnCyan = 46] = "OnCyan";
  t[t.OnWhite = 47] = "OnWhite";
  t[t.OnBackground = 49] = "OnBackground";
  t[t.BrightBlack = 90] = "BrightBlack";
  t[t.BrightRed = 91] = "BrightRed";
  t[t.BrightGreen = 92] = "BrightGreen";
  t[t.BrightYellow = 93] = "BrightYellow";
  t[t.BrightBlue = 94] = "BrightBlue";
  t[t.BrightMagenta = 95] = "BrightMagenta";
  t[t.BrightCyan = 96] = "BrightCyan";
  t[t.BrightWhite = 97] = "BrightWhite";
  t[t.OnBrightBlack = 100] = "OnBrightBlack";
  t[t.OnBrightRed = 101] = "OnBrightRed";
  t[t.OnBrightGreen = 102] = "OnBrightGreen";
  t[t.OnBrightYellow = 103] = "OnBrightYellow";
  t[t.OnBrightBlue = 104] = "OnBrightBlue";
  t[t.OnBrightMagenta = 105] = "OnBrightMagenta";
  t[t.OnBrightCyan = 106] = "OnBrightCyan";
  t[t.OnBrightWhite = 107] = "OnBrightWhite";
  t.DoubleUnderline = "4:2";
  t.CurlyUnderline = "4:3";
  t.DottedUnderline = "4:4";
  t.DashedUnderline = "4:5";
  return t;
}(L || {});

var R = "[";

function cmd(t, r, i) {
  if (!P && t === M.Style) {
    return "";
  }
  var n = R;
  if (t === M.SetPrivateMode) {
    n += "?";
    t = M.SetMode;
  } else if (t === M.UnsetPrivateMode) {
    n += "?";
    t = M.UnsetMode;
  }
  if (Array.isArray(r)) {
    n += r.join(";");
  } else if (null != r) {
    n += `${r}`;
  }
  return n += t;
}

var j = /([\x1B\x9B][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><])/g;

var stripAnsi = t => t.replace(j, "");

class CLIError extends Error {
  constructor(t, r) {
    super(stripAnsi(t));
    this.output = t;
    this.exit = null != r ? r : 1;
  }
  toString() {
    return this.output;
  }
}

function text(t, ...r) {
  var i = "";
  if (Array.isArray(t)) {
    var n = 0;
    for (var a = 0; a < t.length; a++) {
      i += t[a];
      if (n < r.length) {
        i += r[n++];
      }
    }
  } else if ("string" == typeof t) {
    i += t;
    for (var o of r) {
      i += o;
    }
  }
  return i;
}

function error(t, ...r) {
  return "number" == typeof t ? new CLIError(text(...r), t) : new CLIError(text(t, ...r));
}

function commit(t) {
  return make((r => {
    var i;
    var n = pipe(t, onPush((t => {
      i = t;
      r.next(t);
    })), onEnd((() => {
      if ("string" == typeof i && !i.endsWith("\n")) {
        r.next("");
      }
      r.complete();
    })), function subscribe(t) {
      return r => {
        var i = e;
        var n = !1;
        r((r => {
          if (0 === r) {
            n = !0;
          } else if (0 === r.tag) {
            (i = r[0])(0);
          } else if (!n) {
            t(r[0]);
            i(0);
          }
        }));
        return {
          unsubscribe() {
            if (!n) {
              n = !0;
              i(1);
            }
          }
        };
      };
    }((() => {})));
    return n.unsubscribe;
  }));
}

function compose(t) {
  var r = cmd(M.Style, [ L.Reset, L.NoInvert ]);
  var i = pipe(fromAsyncIterable(async function* convertError(t) {
    try {
      yield* t;
    } catch (t) {
      if (t instanceof CLIError) {
        process.exitCode = t.exit;
        yield t;
      } else {
        yield ("" + t).trim();
      }
    }
    yield "\n";
  }(t)), concatMap((t => "object" == typeof t && !(t instanceof CLIError) ? compose(t) : fromValue(t))), filter((t => null != t)), share);
  return pipe(i, concatMap((t => {
    var n = pipe("string" == typeof t || t instanceof CLIError ? fromValue(t) : merge([ t, never ]), takeUntil(i), share);
    return pipe(commit(n), function scan(t, r) {
      return i => n => {
        var a = r;
        i((r => {
          if (0 === r) {
            n(0);
          } else if (0 === r.tag) {
            n(r);
          } else {
            n(push(a = t(a, r[0])));
          }
        }));
      };
    }(((t, i) => "string" == typeof i ? function clear(t) {
      var r = 0;
      for (var i = 0; i < t.length; i++) {
        if (10 === t.charCodeAt(i)) {
          r++;
        }
      }
      if (r) {
        return cmd(M.PrevLine, r) + cmd(M.DeleteLines, r + 1);
      } else if (stripAnsi(t)) {
        return cmd(M.EraseLine, B.Backward) + cmd(M.ToColumn, 1);
      } else {
        return "";
      }
    }("string" == typeof t ? t : "") + i + r : i), ""));
  })), takeUntil(pipe(i, function takeLast(t) {
    return r => i => {
      var n = [];
      var a = e;
      r((r => {
        if (0 === r) {
          T(n)(i);
        } else if (0 === r.tag) {
          (a = r[0])(0);
        } else {
          if (n.length >= t && t) {
            n.shift();
          }
          n.push(r[0]);
          a(0);
        }
      }));
    };
  }(1))));
}

var D = function(t) {
  t.Bell = "";
  t.Newline = "\n";
  t.Tab = "\t";
  t.Space = " ";
  t.Ellipsis = "…";
  return t;
}({});

var _ = function(t) {
  t.TopLeft = "┏";
  t.TopRight = "┓";
  t.BottomLeft = "┗";
  t.BottomRight = "┛";
  t.Vertical = "┃";
  t.VerticalRight = "┣";
  t.VerticalLeft = "┫";
  t.Horizontal = "━";
  t.HorizontalDown = "┳";
  t.HorizontalUp = "┻";
  t.Cross = "╋";
  return t;
}({});

var U = function(t) {
  t.Tick = "✓";
  t.TickSwoosh = "✔";
  t.Cross = "✖";
  t.CrossSwoosh = "✘";
  t.Home = "⌂";
  t.Note = "♪";
  t.Warning = "⚠";
  t.Info = "ℹ";
  t.Star = "★";
  return t;
}({});

var q = [ "⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏" ];

var H = [ "◐", "◓", "◑", "◒" ];

var V = !!process.env.GITHUB_ACTIONS;

var toCommandValue = t => "string" == typeof t || null == t ? t ? "" + t : "" : JSON.stringify(t);

var escapeData = t => toCommandValue(t).replace(/%/g, "%25").replace(/\r/g, "%0D").replace(/\n/g, "%0A");

var escapeProperty = t => toCommandValue(t).replace(/%/g, "%25").replace(/\r/g, "%0D").replace(/\n/g, "%0A").replace(/:/g, "%3A").replace(/,/g, "%2C");

function githubAnnotation(t, r, i) {
  if (V) {
    var n = `::${t}`;
    var a = "";
    if (i) {
      for (var o in i) {
        if (i) {
          a += ",";
        }
        if (i[o]) {
          a += `${o}=${escapeProperty(i[o])}`;
        }
      }
    }
    if (a) {
      n += ` ${a}`;
    }
    n += `::${escapeData(r)}\n`;
    process.stdout.write(n);
  }
}

function initTTY(t = {}) {
  var r = "dumb" !== process.env.TERM && !process.env.CI && !t.disableTTY;
  var i = null;
  var a = process.stdout;
  if (V) {
    a = process.stderr;
  } else if (!a.isTTY && process.stderr.isTTY) {
    a = process.stderr;
    i = process.stdout;
  } else {
    r = a.isTTY;
  }
  var o = process.argv.includes("--color");
  var s = "FORCE_COLOR" in process.env || !process.env.NO_COLOR && !process.env.CI;
  _setColor(r && s || o || V);
  function _terminate(t) {
    !function _restore() {
      if (process.stdin.isTTY) {
        process.stdin.setRawMode(!1);
      }
      a.write(cmd(M.Reset) + cmd(M.ResetPrivateMode) + cmd(M.SetPrivateMode, A.ShowCursor));
    }();
    process.exit(t);
  }
  function _signal(t) {
    _terminate("SIGINT" === t ? 130 : 143);
  }
  var l = pipe(function fromReadStream(t, r) {
    return make((i => {
      function onKeypress(t, n) {
        switch (n.name) {
         case "c":
         case "d":
         case "x":
          if (n.ctrl) {
            return r();
          }
          break;

         case "escape":
          return cleanup();
        }
        i.next({
          ...n,
          data: t
        });
      }
      function cleanup() {
        if (t.isTTY) {
          t.setRawMode(!1);
        }
        i.complete();
        t.removeListener("keypress", onKeypress);
        t.unref();
      }
      if (t.isTTY) {
        t.setRawMode(!0);
      }
      n.emitKeypressEvents(t);
      t.setEncoding("utf8");
      t.resume();
      t.addListener("keypress", onKeypress);
      return cleanup;
    }));
  }(process.stdin, (() => _terminate(130))), function onStart(t) {
    return r => i => r((r => {
      if (0 === r) {
        i(0);
      } else if (0 === r.tag) {
        i(r);
        t();
      } else {
        i(r);
      }
    }));
  }((function _start() {
    _setColor(r && s || o);
    if (r) {
      a.write(cmd(M.UnsetPrivateMode, A.ShowCursor));
      process.on("SIGINT", _signal);
      process.on("SIGTERM", _signal);
    }
  })), onEnd((function _end() {
    if (r) {
      process.removeListener("SIGINT", _signal);
      process.removeListener("SIGTERM", _signal);
      a.write(cmd(M.Reset) + cmd(M.ResetPrivateMode) + cmd(M.SetPrivateMode, A.ShowCursor));
    }
  })), share);
  var c = pipe(concat([ pipe(l, filter((() => !1))), fromValue(null) ]), share);
  return {
    isInteractive: r,
    output: a,
    pipeTo: i,
    inputSource: l,
    cancelSource: c,
    write: function write(...t) {
      a.write(text(...t));
    },
    start: function start(r) {
      var write = t => a.write("" + t);
      if (t.disableTTY) {
        return pipe(compose(r), onPush(write), toPromise);
      } else {
        return pipe(compose(r), onPush(write), takeUntil(c), toPromise);
      }
    },
    mode: function mode(...t) {
      if (r) {
        var i = [];
        var n = [];
        for (var o of t) {
          if (o === F.Insert || o === F.AutomaticNewline) {
            i.push(o);
          } else {
            n.push(o);
          }
        }
        if (i.length) {
          a.write(cmd(M.SetMode, i));
        }
        if (n.length) {
          a.write(cmd(M.SetPrivateMode, n));
        }
      }
    },
    modeOff: function modeOff(...t) {
      if (r) {
        var i = [];
        var n = [];
        for (var o of t) {
          if (o === F.Insert || o === F.AutomaticNewline) {
            i.push(o);
          } else {
            n.push(o);
          }
        }
        if (i.length) {
          a.write(cmd(M.UnsetMode, i));
        }
        if (n.length) {
          a.write(cmd(M.UnsetPrivateMode, n));
        }
      }
    }
  };
}

function indent(t, r) {
  if (t.includes("\n")) {
    var i = t.trim().split("\n").join(text([ D.Newline, r ]));
    return t.endsWith("\n") ? i + "\n" : i;
  } else {
    return t;
  }
}

function code(t) {
  return text`${cmd(M.Style, L.Underline)}${t}${cmd(M.Style, L.NoUnderline)}`;
}

function bold(t) {
  return text`${cmd(M.Style, L.Bold)}${t}${cmd(M.Style, L.Normal)}`;
}

function hint(t) {
  return text([ cmd(M.Style, L.BrightBlack), `${_.BottomLeft} `, cmd(M.Style, L.BrightBlue), `${U.Info} `, cmd(M.Style, L.Blue), indent(t, "    ") ]);
}

function errorMessage$1(t) {
  return error([ "\n", cmd(M.Style, [ L.Red, L.Invert ]), ` ${U.Warning} Error `, cmd(M.Style, L.NoInvert), `\n${t.trim()}\n` ]);
}

function externalError(t, r) {
  var i;
  var n;
  if (r && "object" == typeof r) {
    if ("name" in r && ("TSError" === r.name || "TadaError" === r.name || "code" in r)) {
      i = "code" in r ? "System Error" : "Error";
      n = "TadaError" === r.name ? text([ cmd(M.Style, L.Blue), r.message ]) : r.message.trim();
    } else if ("stack" in r && "string" == typeof r.stack) {
      i = "Unexpected Error";
      n = `${r.stack}`;
    } else if ("message" in r && "string" == typeof r.message) {
      i = "Unexpected Error";
      n = `${r.message}`;
    } else {
      i = "Unexpected Error";
      n = `${r}`;
    }
  } else {
    i = "Unexpected Error";
    n = `${r}`;
  }
  return error([ "\n", cmd(M.Style, [ L.Red, L.Invert ]), ` ${U.Warning} ${i} `, cmd(M.Style, L.NoInvert), `\n${t.trim()}\n`, cmd(M.Style, L.BrightBlack), `${_.BottomLeft} `, indent(!n.endsWith("\n") ? n + "\n" : n, "  ") ]);
}

function experimentMessage(t) {
  return error([ cmd(M.Style, [ L.Yellow, L.Bold ]), `${U.Warning} Note: `, cmd(M.Style, L.Reset), `${t.trim()}\n\n` ]);
}

var directoryExists = async t => {
  try {
    var r = await u.stat(t);
    if (r.isDirectory()) {
      return !0;
    } else if (r.isSymbolicLink()) {
      return directoryExists(await u.realpath(t));
    } else {
      return !1;
    }
  } catch {
    return !1;
  }
};

var writeOutput = async (r, i) => {
  if (r && "object" == typeof r && "writable" in r) {
    return await new Promise(((t, n) => {
      r.write(i, (r => {
        if (r) {
          n(r);
        } else {
          t();
        }
      }));
    }));
  }
  var n = t.dirname("string" != typeof r ? await u.realpath(r) : r);
  if (!await directoryExists(n)) {
    await u.mkdir(n, {
      recursive: !0
    });
  }
  if (!await (o = r, u.stat(o).then((t => t.isFile())).catch((() => !1)))) {
    await u.writeFile(r, i);
  } else {
    var a = r + ".tmp";
    await u.writeFile(a, i);
    try {
      await u.rename(a, r);
    } catch (t) {
      await u.unlink(a);
      throw t;
    } finally {
      await (async t => {
        try {
          var r = new Date;
          await u.utimes(t, r, r);
        } catch (t) {}
      })(r);
    }
  }
  var o;
};

function summary$1(t) {
  var r = text([ cmd(M.Style, L.BrightGreen), `${U.Tick} Introspection output was generated successfully\n` ]);
  if (t) {
    r += hint(`The pipe output was generated in the ${code(".d.ts")} format.\nFor the ${code(".ts")} format, pass the ${code("--force-ts-format")} argument.\n`);
  }
  return r;
}

async function* run$5(t, i) {
  var n;
  var a;
  try {
    n = await r.loadConfig(i.tsconfig);
    a = r.parseConfig(n.pluginConfig, n.rootPath);
  } catch (t) {
    throw externalError("Failed to load configuration.", t);
  }
  var o = r.loadRef(a);
  try {
    o = await o.load({
      rootPath: c.dirname(n.configPath)
    });
  } catch (t) {
    throw externalError("Failed to load schema(s).", t);
  }
  var s = c.dirname(n.configPath);
  if ("schema" in a) {
    var l = o.current;
    var u;
    if (!i.output && t.pipeTo) {
      u = t.pipeTo;
    } else if (i.output) {
      u = c.resolve(process.cwd(), i.output);
    } else if (a.tadaOutputLocation) {
      u = a.tadaOutputLocation;
    } else {
      throw errorMessage$1("No output path was specified to write the output file to.\n" + hint(`You have to either set ${code('"tadaOutputLocation"')} in your configuration,\npass an ${code("--output")} argument to this command,\nor pipe this command to an output file.`));
    }
    var d;
    try {
      d = r.outputIntrospectionFile(r.minifyIntrospection(l.introspection), {
        fileType: u && "string" == typeof u ? u : i.forceTSFormat ? ".ts" : ".d.ts",
        shouldPreprocess: !i.disablePreprocessing
      });
    } catch (t) {
      throw externalError("Could not generate introspection output", t);
    }
    try {
      await writeOutput(u, d);
    } catch (t) {
      throw externalError("Something went wrong while writing the introspection file", t);
    }
    yield summary$1(!i.forceTSFormat && "string" != typeof u);
  } else {
    if (i.output) {
      throw errorMessage$1("Output path was specified, while multiple schemas are configured.\n" + hint(`You can only output all schemas to their ${code('"tadaOutputLocation"')} options\nwhen multiple ${code("schemas")} are set up.`));
    }
    for (var f in o.multi) {
      var g = o.multi[f];
      if (!g) {
        continue;
      } else if (!g.tadaOutputLocation) {
        throw errorMessage$1(`No output path was specified to write the '${f}' schema to.\n` + hint(`You have to set ${code('"tadaOutputLocation"')} in each schema configuration.`));
      }
      var p = void 0;
      try {
        p = r.outputIntrospectionFile(r.minifyIntrospection(g.introspection), {
          fileType: g.tadaOutputLocation,
          shouldPreprocess: !i.disablePreprocessing
        });
      } catch (t) {
        throw externalError(`Could not generate any output for the '${f}' schema`, t);
      }
      try {
        await writeOutput(c.resolve(s, g.tadaOutputLocation), p);
      } catch (t) {
        throw externalError(`Something went wrong while writing the '${f}' schema's output`, t);
      }
    }
    yield summary$1();
  }
}

class GenerateOutputCommand extends Command {
  static paths=[ [ "generate-output" ], [ "generate", "output" ] ];
  forceTSFormat=Boolean("--force-ts-format", !1, {
    description: "Forces the `.ts` output format when the output is piped",
    hidden: !0
  });
  disablePreprocessing=Boolean("--disable-preprocessing", !1, {
    description: "Disables pre-processing, which is an internal introspection format generated ahead of time"
  });
  tsconfig=String$1("--tsconfig,-c", {
    description: "Specify the `tsconfig.json` used to read, unless `--output` is passed."
  });
  output=String$1("--output,-o", {
    description: "Specifies where to output the file to.\tDefault: The `tadaOutputLocation` configuration option"
  });
  async execute() {
    var t = initTTY();
    var r = await t.start(run$5(t, {
      forceTSFormat: this.forceTSFormat,
      disablePreprocessing: this.disablePreprocessing,
      output: this.output,
      tsconfig: this.tsconfig
    }));
    return exitCode() || ("object" == typeof r ? r.exit : 0);
  }
}

async function generateOutput(t) {
  var r = initTTY({
    disableTTY: !0
  });
  var i = await r.start(run$5(r, t));
  if (i instanceof Error) {
    throw i;
  }
}

var W = process.cwd();

var K = "  ";

function warningFile$1(t) {
  var r = c.relative(W, t);
  if (!r.startsWith("..")) {
    t = r;
  }
  return text([ cmd(M.Style, L.Underline), t, cmd(M.Style, L.NoUnderline), "\n" ]);
}

var documentSummary$1 = t => {
  var r = "";
  if ("number" != typeof t ? Object.values(t).every((t => !t)) : !t) {
    r += text([ cmd(M.Style, L.Blue), `${U.Info} No persisted documents were found `, cmd(M.Style, L.BrightBlack), "(Persisted manifests were not generated)\n" ]);
  } else if ("number" == typeof t) {
    r += text([ cmd(M.Style, L.BrightGreen), `${U.Tick} Persisted manifest was generated successfully `, cmd(M.Style, L.BrightBlack), `(${t} documents extracted)\n` ]);
  } else {
    r += text([ cmd(M.Style, L.BrightGreen), `${U.Tick} Persisted manifests were generated successfully.\n` ]);
    for (var i in t) {
      r += text([ cmd(M.Style, L.BrightBlack), `${_.BottomLeft} `, cmd(M.Style, L.BrightBlue), `${t[i]} documents extracted for the '${i}' schema\n` ]);
    }
  }
  return r;
};

function warningSummary$1(t) {
  return error([ cmd(M.Style, L.Red), `${U.Cross} ${t} warnings\n` ]);
}

function infoSummary$2(t, r) {
  var i = "";
  if (t) {
    i += text([ cmd(M.Style, L.BrightYellow), U.Warning, ` ${t} warnings\n` ]);
  }
  return i += documentSummary$1(r);
}

function warningGithub$1(t) {
  githubAnnotation("warning", t.message, {
    file: t.file,
    line: t.line,
    col: t.col
  });
}

function runningPersisted(t, r) {
  var i = t ? r ? `(${t}/${r})` : `(${t})` : "";
  var frame = t => text([ cmd(M.Style, L.Magenta), q[t % q.length], " ", cmd(M.Style, L.Foreground), `Scanning files${D.Ellipsis} `, cmd(M.Style, L.BrightBlack), i ]);
  return concat([ fromValue(frame(0)), pipe(interval(150), map((t => frame(t + 1)))) ]);
}

async function* run$4(t, i) {
  var {runPersisted: n} = await Promise.resolve().then((function() {
    return require("./thread-chunk.js");
  }));
  var a;
  var o;
  try {
    a = await r.loadConfig(i.tsconfig);
    o = r.parseConfig(a.pluginConfig, a.rootPath);
  } catch (t) {
    throw externalError("Failed to load configuration.", t);
  }
  if (t.isInteractive) {
    yield runningPersisted();
  }
  var s = n({
    disableNormalization: !!i.disableNormalization,
    rootPath: a.rootPath,
    configPath: a.configPath,
    pluginConfig: o
  });
  var l = [];
  var u = 0;
  var d = 0;
  var f = 0;
  try {
    if (t.isInteractive) {
      yield runningPersisted();
    }
    for await (var g of s) {
      if ("EXTERNAL_WARNING" === g.kind) {
        yield experimentMessage(`${code(".vue")} and ${code(".svelte")} file support is experimental.`);
      } else if ("FILE_COUNT" === g.kind) {
        d = g.fileCount;
      } else {
        f++;
        l.push(...g.documents);
        u += g.warnings.length;
        if (g.warnings.length) {
          var p = warningFile$1(g.filePath);
          for (var h of g.warnings) {
            p += (v = h, text([ K, cmd(M.Style, L.BrightBlack), `${v.line}:${v.col}`, D.Tab, cmd(M.Style, L.Foreground), indent(v.message.trim(), text([ K, D.Tab ])), D.Newline ]));
            warningGithub$1(h);
          }
          yield p + "\n";
        }
      }
      if (t.isInteractive) {
        yield runningPersisted(f, d);
      }
    }
  } catch (t) {
    throw externalError("Could not generate persisted manifest file", t);
  }
  var v;
  var m = c.dirname(a.configPath);
  if ("schema" in o) {
    var y;
    if (!i.output && t.pipeTo) {
      y = t.pipeTo;
    } else if (i.output) {
      y = c.resolve(process.cwd(), i.output);
    } else if (o.tadaPersistedLocation) {
      y = c.resolve(c.dirname(a.configPath), o.tadaPersistedLocation);
    } else {
      throw errorMessage$1("No output path was specified to write the persisted manifest file to.\n" + hint(`You have to either set ${code('"tadaPersistedLocation"')} in your configuration,\npass an ${code("--output")} argument to this command,\nor pipe this command to an output file.`));
    }
    if (u && i.failOnWarn) {
      throw warningSummary$1(u);
    } else if (l.length) {
      try {
        var S = {};
        for (var w of l) {
          S[w.hashKey] = w.document;
        }
        var b = JSON.stringify(S, null, 2);
        await writeOutput(y, b);
      } catch (t) {
        throw externalError("Something went wrong while writing the persisted manifest file.", t);
      }
    }
    yield infoSummary$2(u, l.length);
  } else {
    if (i.output) {
      throw errorMessage$1("Output path was specified, while multiple schemas are configured.\n" + hint(`You can only output all schemas to their ${code('"tadaPersistedLocation"')} options\nwhen multiple ${code("schemas")} are set up.`));
    }
    var x = {};
    for (var E of o.schemas) {
      var {name: C, tadaPersistedLocation: O} = E;
      if (!O) {
        throw errorMessage$1(`No output path was specified to write the '${C}' schema to.\n` + hint(`You have to set ${code('"tadaPersistedLocation"')} in each schema configuration.`));
      }
      try {
        x[C] = 0;
        var $ = {};
        for (var k of l) {
          if (k.schemaName === C) {
            $[k.hashKey] = k.document;
            x[C]++;
          }
        }
        if (x[C]) {
          var N = JSON.stringify($, null, 2);
          await writeOutput(c.resolve(m, O), N);
        }
      } catch (t) {
        throw externalError(`Something went wrong while writing the '${C}' schema's persisted manifest file.`, t);
      }
    }
    if (u && i.failOnWarn) {
      throw warningSummary$1(u);
    } else {
      yield infoSummary$2(u, x);
    }
  }
}

class GeneratePersisted extends Command {
  static paths=[ [ "generate-persisted" ], [ "generate", "persisted" ] ];
  tsconfig=String$1("--tsconfig,-c", {
    description: "Specify the `tsconfig.json` used to read, unless `--output` is passed."
  });
  disableNormalization=Boolean("--disable-normalization", !1, {
    description: "Disables normalizing of GraphQL documents (parsing then printing documents)"
  });
  failOnWarn=Boolean("--fail-on-warn", !1, {
    description: "Triggers an error and a non-zero exit code if any warnings have been reported"
  });
  output=String$1("--output,-o", {
    description: "Specifies where to output the file to.\tDefault: The `tadaPersistedLocation` configuration option"
  });
  async execute() {
    var t = initTTY();
    var r = await t.start(run$4(t, {
      disableNormalization: this.disableNormalization,
      failOnWarn: this.failOnWarn,
      output: this.output,
      tsconfig: this.tsconfig
    }));
    return exitCode() || ("object" == typeof r ? r.exit : 0);
  }
}

async function generatePersisted(t) {
  var r = initTTY({
    disableTTY: !0
  });
  var i = await r.start(run$4(r, t));
  if (i instanceof Error) {
    throw i;
  }
}

var G = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

function getPrintable(t) {
  if (null === t) {
    return "null";
  }
  if (void 0 === t) {
    return "undefined";
  }
  if ("" === t) {
    return "an empty string";
  }
  if ("symbol" == typeof t) {
    return `<${t.toString()}>`;
  }
  if (Array.isArray(t)) {
    return "an array";
  }
  return JSON.stringify(t);
}

function computeKey(t, r) {
  var i, n, a;
  if ("number" == typeof r) {
    return `${null !== (i = null == t ? void 0 : t.p) && void 0 !== i ? i : "."}[${r}]`;
  } else if (G.test(r)) {
    return `${null !== (n = null == t ? void 0 : t.p) && void 0 !== n ? n : ""}.${r}`;
  } else {
    return `${null !== (a = null == t ? void 0 : t.p) && void 0 !== a ? a : "."}[${JSON.stringify(r)}]`;
  }
}

function pushError({errors: t, p: r} = {}, i) {
  null == t || t.push(`${null != r ? r : "."}: ${i}`);
  return !1;
}

function makeCoercionFn(t, r) {
  return i => {
    var n = t[r];
    t[r] = i;
    return makeCoercionFn(t, r).bind(null, n);
  };
}

function isLiteral(t) {
  return makeValidator({
    test: (r, i) => {
      if (r !== t) {
        return pushError(i, `Expected ${getPrintable(t)} (got ${getPrintable(r)})`);
      }
      return !0;
    }
  });
}

function isString() {
  return makeValidator({
    test: (t, r) => {
      if ("string" != typeof t) {
        return pushError(r, `Expected a string (got ${getPrintable(t)})`);
      }
      return !0;
    }
  });
}

function isArray(t, {delimiter: r} = {}) {
  return makeValidator({
    test: (i, n) => {
      var a;
      var o = i;
      if ("string" == typeof i && void 0 !== r) {
        if (void 0 !== (null == n ? void 0 : n.coercions)) {
          if (void 0 === (null == n ? void 0 : n.coercion)) {
            return pushError(n, "Unbound coercion result");
          }
          i = i.split(r);
        }
      }
      if (!Array.isArray(i)) {
        return pushError(n, `Expected an array (got ${getPrintable(i)})`);
      }
      var s = !0;
      for (var l = 0, c = i.length; l < c; ++l) {
        if (!(s = t(i[l], Object.assign(Object.assign({}, n), {
          p: computeKey(n, l),
          coercion: makeCoercionFn(i, l)
        })) && s) && null == (null == n ? void 0 : n.errors)) {
          break;
        }
      }
      if (i !== o) {
        n.coercions.push([ null !== (a = n.p) && void 0 !== a ? a : ".", n.coercion.bind(null, i) ]);
      }
      return s;
    }
  });
}

function isTuple(t, {delimiter: r} = {}) {
  var i = hasExactLength(t.length);
  return makeValidator({
    test: (n, a) => {
      var o;
      if ("string" == typeof n && void 0 !== r) {
        if (void 0 !== (null == a ? void 0 : a.coercions)) {
          if (void 0 === (null == a ? void 0 : a.coercion)) {
            return pushError(a, "Unbound coercion result");
          }
          n = n.split(r);
          a.coercions.push([ null !== (o = a.p) && void 0 !== o ? o : ".", a.coercion.bind(null, n) ]);
        }
      }
      if (!Array.isArray(n)) {
        return pushError(a, `Expected a tuple (got ${getPrintable(n)})`);
      }
      var s = i(n, Object.assign({}, a));
      for (var l = 0, c = n.length; l < c && l < t.length; ++l) {
        if (!(s = t[l](n[l], Object.assign(Object.assign({}, a), {
          p: computeKey(a, l),
          coercion: makeCoercionFn(n, l)
        })) && s) && null == (null == a ? void 0 : a.errors)) {
          break;
        }
      }
      return s;
    }
  });
}

function isRecord(t, {keys: r = null} = {}) {
  var i = isArray(isTuple([ null != r ? r : isString(), t ]));
  return makeValidator({
    test: (n, a) => {
      var o;
      if (Array.isArray(n)) {
        if (void 0 !== (null == a ? void 0 : a.coercions)) {
          if (void 0 === (null == a ? void 0 : a.coercion)) {
            return pushError(a, "Unbound coercion result");
          }
          if (!i(n, Object.assign(Object.assign({}, a), {
            coercion: void 0
          }))) {
            return !1;
          }
          n = Object.fromEntries(n);
          a.coercions.push([ null !== (o = a.p) && void 0 !== o ? o : ".", a.coercion.bind(null, n) ]);
          return !0;
        }
      }
      if ("object" != typeof n || null === n) {
        return pushError(a, `Expected an object (got ${getPrintable(n)})`);
      }
      var s = Object.keys(n);
      var l = !0;
      for (var c = 0, u = s.length; c < u && (l || null != (null == a ? void 0 : a.errors)); ++c) {
        var d = s[c];
        var f = n[d];
        if ("__proto__" === d || "constructor" === d) {
          l = pushError(Object.assign(Object.assign({}, a), {
            p: computeKey(a, d)
          }), "Unsafe property name");
          continue;
        }
        if (null !== r && !r(d, a)) {
          l = !1;
          continue;
        }
        if (!t(f, Object.assign(Object.assign({}, a), {
          p: computeKey(a, d),
          coercion: makeCoercionFn(n, d)
        }))) {
          l = !1;
          continue;
        }
      }
      return l;
    }
  });
}

var isOneOf = (t, {exclusive: r = !1} = {}) => makeValidator({
  test: (i, n) => {
    var a, o, s;
    var l = [];
    var c = void 0 !== (null == n ? void 0 : n.errors) ? [] : void 0;
    for (var u = 0, d = t.length; u < d; ++u) {
      var f = void 0 !== (null == n ? void 0 : n.errors) ? [] : void 0;
      var g = void 0 !== (null == n ? void 0 : n.coercions) ? [] : void 0;
      if (t[u](i, Object.assign(Object.assign({}, n), {
        errors: f,
        coercions: g,
        p: `${null !== (a = null == n ? void 0 : n.p) && void 0 !== a ? a : "."}#${u + 1}`
      }))) {
        l.push([ `#${u + 1}`, g ]);
        if (!r) {
          break;
        }
      } else {
        null == c || c.push(f[0]);
      }
    }
    if (1 === l.length) {
      var [, p] = l[0];
      if (void 0 !== p) {
        null === (o = null == n ? void 0 : n.coercions) || void 0 === o || o.push(...p);
      }
      return !0;
    }
    if (l.length > 1) {
      pushError(n, `Expected to match exactly a single predicate (matched ${l.join(", ")})`);
    } else {
      null === (s = null == n ? void 0 : n.errors) || void 0 === s || s.push(...c);
    }
    return !1;
  }
});

function makeTrait(t) {
  return () => t;
}

function makeValidator({test: t}) {
  return makeTrait(t)();
}

function hasExactLength(t) {
  return makeValidator({
    test: (r, i) => {
      if (r.length !== t) {
        return pushError(i, `Expected to have a length of exactly ${t} elements (got ${r.length})`);
      }
      return !0;
    }
  });
}

function matchesRegExp(t) {
  return makeValidator({
    test: (r, i) => {
      if (!t.test(r)) {
        return pushError(i, `Expected to match the pattern ${t.toString()} (got ${getPrintable(r)})`);
      }
      return !0;
    }
  });
}

function cascade(t, ...r) {
  var i = Array.isArray(r[0]) ? r[0] : r;
  return makeValidator({
    test: (r, n) => {
      var a, o;
      var s = {
        value: r
      };
      var l = void 0 !== (null == n ? void 0 : n.coercions) ? makeCoercionFn(s, "value") : void 0;
      var c = void 0 !== (null == n ? void 0 : n.coercions) ? [] : void 0;
      if (!t(r, Object.assign(Object.assign({}, n), {
        coercion: l,
        coercions: c
      }))) {
        return !1;
      }
      var u = [];
      if (void 0 !== c) {
        for (var [, d] of c) {
          u.push(d());
        }
      }
      try {
        if (void 0 !== (null == n ? void 0 : n.coercions)) {
          if (s.value !== r) {
            if (void 0 === (null == n ? void 0 : n.coercion)) {
              return pushError(n, "Unbound coercion result");
            }
            n.coercions.push([ null !== (a = n.p) && void 0 !== a ? a : ".", n.coercion.bind(null, s.value) ]);
          }
          null === (o = null == n ? void 0 : n.coercions) || void 0 === o || o.push(...c);
        }
        return i.every((t => t(s.value, n)));
      } finally {
        for (var f of u) {
          f();
        }
      }
    }
  });
}

var Y;

!function(t) {
  t.Forbids = "Forbids";
  t.Requires = "Requires";
}(Y || (Y = {}));

var J = {
  __proto__: null,
  get KeyRelationship() {
    return Y;
  },
  applyCascade: function applyCascade(t, ...r) {
    return cascade(t, Array.isArray(r[0]) ? r[0] : r);
  },
  cascade,
  hasExactLength,
  isArray,
  isDict: function isDict(t, r = {}) {
    return isRecord(t, r);
  },
  isLiteral,
  isOneOf,
  isRecord,
  isString,
  isTuple,
  isUnknown: function isUnknown() {
    return makeValidator({
      test: (t, r) => !0
    });
  },
  makeTrait,
  makeValidator,
  matchesRegExp
};

async function* run$3(i, n) {
  var o = n.headers ? {
    url: n.input,
    headers: n.headers
  } : n.input;
  var s = r.load({
    rootPath: process.cwd(),
    origin: o
  });
  var l;
  try {
    l = (await s.load()).schema;
  } catch (t) {
    throw externalError("Failed to load schema.", t);
  }
  var c;
  if (!n.output && i.pipeTo) {
    c = i.pipeTo;
  } else if (n.output) {
    c = t.resolve(process.cwd(), n.output);
  } else {
    var u;
    var d;
    try {
      u = await r.loadConfig(n.tsconfig);
      d = r.parseConfig(u.pluginConfig, u.rootPath);
    } catch (t) {
      throw externalError("Failed to load configuration.", t);
    }
    if ("schema" in d && "string" == typeof d.schema && ".graphql" === t.extname(d.schema)) {
      c = t.resolve(t.dirname(u.configPath), d.schema);
    } else if (!("schema" in d)) {
      throw errorMessage$1(`Output path cannot be automatically determined when multiple schemas are configured,\nbecause multiple ${code("schemas")} are set up.` + hint(`You have to explicitly pass an ${code("--output")} argument to this command,\nor pipe this command to an output file.`));
    } else {
      throw errorMessage$1(`Output path cannot be automatically determined,\nbecause ${code("schema")} is not a file path.\n` + hint(`You have to either set ${code('"schema"')} in your configuration to a ${code(".graphql")} file,\npass an ${code("--output")} argument to this command,\nor pipe this command to an output file.`));
    }
  }
  try {
    await writeOutput(c, a.printSchema(l));
  } catch (t) {
    throw externalError("Something went wrong while writing the introspection file", t);
  }
  yield function summary() {
    return text([ cmd(M.Style, L.BrightGreen), `${U.Tick} Schema was generated successfully\n` ]);
  }();
}

var z = isArray(cascade(isString(), matchesRegExp(/^[\w-]+[ ]*:[ ]*.+/i)));

var parseHeaders = t => {
  if (t && t.length) {
    return (t || []).reduce(((t, r) => {
      var i = r.indexOf(":");
      var n = r.slice(0, i);
      var a = r.slice(i + 1);
      t[n.trimEnd()] = a.trimStart();
      return t;
    }), {});
  }
};

class GenerateSchema extends Command {
  static paths=[ [ "generate-schema" ], [ "generate", "schema" ] ];
  input=String$1({
    name: "schema",
    required: !0
  });
  tsconfig=String$1("--tsconfig,-c", {
    description: "Specify the `tsconfig.json` used to read, unless `--output` is passed."
  });
  output=String$1("--output,-o", {
    description: "Specify where to output the file to.\tDefault: The `schema` configuration option, if it's a file path"
  });
  headers=function Array$1(t, r, i) {
    var [n, a] = rerouteArguments(r, {});
    var {arity: o = 1} = a;
    var s = t.split(",");
    var l = new Set(s);
    return makeCommandOption({
      definition(t) {
        t.addOption({
          names: s,
          arity: o,
          hidden: null == a ? void 0 : a.hidden,
          description: null == a ? void 0 : a.description,
          required: a.required
        });
      },
      transformer(t, r, i) {
        var o;
        var s = void 0 !== n ? [ ...n ] : void 0;
        for (var {name: c, value: u} of i.options) {
          if (!l.has(c)) {
            continue;
          }
          o = c;
          (s = null != s ? s : []).push(u);
        }
        if (void 0 !== s) {
          return applyValidator(null != o ? o : r, s, a.validator);
        } else {
          return s;
        }
      }
    });
  }("--header", {
    description: "Headers to be used when introspection a schema from a remote URL",
    validator: z
  });
  async execute() {
    var t = initTTY();
    var r = await t.start(run$3(t, {
      input: this.input,
      headers: parseHeaders(this.headers),
      output: this.output,
      tsconfig: this.tsconfig
    }));
    return exitCode() || ("object" == typeof r ? r.exit : 0);
  }
}

async function generateSchema(t) {
  var r = initTTY({
    disableTTY: !0
  });
  var i = await r.start(run$3(r, t));
  if (i instanceof Error) {
    throw i;
  }
}

var X = process.cwd();

var Q = "  ";

function warningFile(t) {
  var r = c.relative(X, t);
  if (!r.startsWith("..")) {
    t = r;
  }
  return text([ cmd(M.Style, L.Underline), t, cmd(M.Style, L.NoUnderline), "\n" ]);
}

var documentSummary = t => {
  var r = "";
  if ("number" == typeof t) {
    r += text([ cmd(M.Style, L.BrightGreen), `${U.Tick} Type cache was generated successfully `, cmd(M.Style, L.BrightBlack), `(${t} document types cached)\n` ]);
  } else {
    r += text([ cmd(M.Style, L.BrightGreen), `${U.Tick} Type caches were generated successfully.\n` ]);
    for (var i in t) {
      r += text([ cmd(M.Style, L.BrightBlack), `${_.BottomLeft} `, cmd(M.Style, L.BrightBlue), `${t[i]} document types cached for the '${i}' schema\n` ]);
    }
  }
  return r;
};

function warningSummary(t) {
  return error([ cmd(M.Style, L.Red), `${U.Cross} ${t} warnings\n` ]);
}

function infoSummary$1(t, r) {
  var i = "";
  if (t) {
    i += text([ cmd(M.Style, L.BrightYellow), U.Warning, ` ${t} warnings\n` ]);
  }
  return i += documentSummary(r);
}

function warningGithub(t) {
  githubAnnotation("warning", t.message, {
    file: t.file,
    line: t.line,
    col: t.col
  });
}

function runningTurbo(t, r) {
  var i = t ? r ? `(${t}/${r})` : `(${t})` : "";
  var frame = t => text([ cmd(M.Style, L.Magenta), q[t % q.length], " ", cmd(M.Style, L.Foreground), `Scanning files${D.Ellipsis} `, cmd(M.Style, L.BrightBlack), i ]);
  return concat([ fromValue(frame(0)), pipe(interval(150), map((t => frame(t + 1)))) ]);
}

var Z = [ "/* eslint-disable */", "/* prettier-ignore */" ].join("\n") + "\n";

async function* run$2(t, i) {
  var {runTurbo: n} = await Promise.resolve().then((function() {
    return require("./thread-chunk2.js");
  }));
  var a;
  var o;
  try {
    a = await r.loadConfig(i.tsconfig);
    o = r.parseConfig(a.pluginConfig, a.rootPath);
  } catch (t) {
    throw externalError("Failed to load configuration.", t);
  }
  var s = c.dirname(a.configPath);
  var l;
  var u = [];
  if ("schema" in o) {
    if (!i.output && t.pipeTo) {
      l = t.pipeTo;
    } else if (i.output) {
      l = c.resolve(process.cwd(), i.output);
    } else if (o.tadaTurboLocation) {
      l = c.resolve(s, o.tadaTurboLocation);
    } else if (o.tadaOutputLocation) {
      l = c.resolve(s, o.tadaOutputLocation, "..", "graphql-cache.d.ts");
      yield function hintMessage$1(t) {
        return error([ cmd(M.Style, [ L.Yellow, L.Bold ]), `${U.Warning} Note: `, cmd(M.Style, L.Reset), `${t.trim()}\n\n` ]);
      }(`No output location was specified.\nThe turbo cache will by default be saved as ${code('"graphql-cache.d.ts"')}.\n` + hint(`To change this, add a ${code('"tadaTurboLocation"')} in your configuration,\npass an ${code("--output")} argument to this command,\nor pipe this command to an output file.`));
    } else {
      throw errorMessage$1("No output path was specified to write the output file to.\n" + hint(`You have to either set ${code('"tadaTurboLocation"')} in your configuration,\npass an ${code("--output")} argument to this command,\nor pipe this command to an output file.`));
    }
  } else if ("schemas" in o) {
    for (var d of o.schemas) {
      if (d.tadaTurboLocation) {
        u.push({
          path: c.resolve(s, d.tadaTurboLocation),
          schemaName: d.name
        });
      }
    }
  }
  var f = n({
    rootPath: a.rootPath,
    configPath: a.configPath,
    pluginConfig: o,
    turboOutputPath: "string" == typeof l ? l : u
  });
  var g = [];
  var p = [];
  var h = 0;
  var v = 0;
  var m = 0;
  try {
    if (t.isInteractive) {
      yield runningTurbo();
    }
    for await (var y of f) {
      if ("EXTERNAL_WARNING" === y.kind) {
        yield experimentMessage(`${code(".vue")} and ${code(".svelte")} file support is experimental.`);
      } else if ("FILE_COUNT" === y.kind) {
        v = y.fileCount;
      } else if ("GRAPHQL_SOURCES" === y.kind) {
        p = y.sources;
      } else {
        m++;
        g.push(...y.documents);
        h += y.warnings.length;
        if (y.warnings.length) {
          var S = warningFile(y.filePath);
          for (var w of y.warnings) {
            S += (b = w, text([ Q, cmd(M.Style, L.BrightBlack), `${b.line}:${b.col}`, D.Tab, cmd(M.Style, L.Foreground), indent(b.message.trim(), text([ Q, D.Tab ])), D.Newline ]));
            warningGithub(w);
          }
          yield S + "\n";
        }
      }
      if (t.isInteractive) {
        yield runningTurbo(m, v);
      }
    }
  } catch (t) {
    throw externalError("Could not build cache", t);
  }
  var b;
  if ("schema" in o) {
    if (h && i.failOnWarn) {
      throw warningSummary(h);
    }
    try {
      var x = {};
      for (var E of g) {
        x[E.argumentKey] = E.documentType;
      }
      var C = createCacheContents(x, p, l);
      await writeOutput(l, C);
    } catch (t) {
      throw externalError("Something went wrong while writing the type cache file", t);
    }
    yield infoSummary$1(h, g.length);
  } else {
    if (i.output) {
      throw errorMessage$1("Output path was specified, while multiple schemas are configured.\n" + hint(`You can only output all schemas to their ${code('"tadaTurboLocation"')} options\nwhen multiple ${code("schemas")} are set up.`));
    }
    var O = {};
    for (var $ of o.schemas) {
      var {name: k, tadaTurboLocation: N} = $;
      if (!N) {
        throw errorMessage$1(`No output path was specified to write the '${k}' type cache to.\n` + hint(`You have to set ${code('"tadaTurboLocation"')} in each schema configuration.`));
      }
      try {
        O[k] = 0;
        var I = {};
        for (var T of g) {
          if (T.schemaName === k) {
            I[T.argumentKey] = T.documentType;
            O[k]++;
          }
        }
        var P = c.resolve(s, N);
        var B = createCacheContents(I, p, P);
        await writeOutput(P, B);
      } catch (t) {
        throw externalError(`Something went wrong while writing the '${k}' schema's type cache file.`, t);
      }
    }
    if (h && i.failOnWarn) {
      throw warningSummary(h);
    } else {
      yield infoSummary$1(h, O);
    }
  }
}

function createCacheContents(t, r, i) {
  var n = "";
  for (var a in t) {
    if (n) {
      n += "\n";
    }
    n += `    ${a}:\n      ${t[a]};`;
  }
  var o = "import type { TadaDocumentNode, $tada } from 'gql.tada';\n";
  var s = "string" == typeof i || i && "object" == typeof i && "toString" in i && !("writable" in i);
  var l = new Set;
  for (var u of r) {
    for (var d of u.imports) {
      if (s) {
        var f = i.toString();
        var g = c.dirname(u.absolutePath);
        if (c.resolve(g, d.specifier) === c.resolve(f) || l.has(d.specifier)) {
          continue;
        }
        l.add(d.specifier);
      }
      o += d.importClause + "\n";
    }
  }
  return Z + o + "\ndeclare module 'gql.tada' {\n interface setupCache {\n" + n + "\n  }\n}\n";
}

class TurboCommand extends Command {
  static paths=[ [ "generate", "turbo" ], [ "turbo" ] ];
  tsconfig=String$1("--tsconfig,-c", {
    description: "Specify the `tsconfig.json` read for configuration."
  });
  failOnWarn=Boolean("--fail-on-warn,-w", !1, {
    description: "Triggers an error and a non-zero exit code if any warnings have been reported"
  });
  output=String$1("--output,-o", {
    description: "Specifies where to output the file to.\tDefault: The `tadaTurboLocation` configuration option"
  });
  async execute() {
    var t = initTTY();
    var r = await t.start(run$2(t, {
      failOnWarn: this.failOnWarn,
      output: this.output,
      tsconfig: this.tsconfig
    }));
    return exitCode() || ("object" == typeof r ? r.exit : 0);
  }
}

async function generateTurbo(t) {
  var r = initTTY({
    disableTTY: !0
  });
  var i = await r.start(run$2(r, t));
  if (i instanceof Error) {
    throw i;
  }
}

var ee = {
  __proto__: null,
  generateOutput,
  generatePersisted,
  generateSchema,
  generateTurbo
};

var te = process.cwd();

var re = "  ";

function diagnosticFile(t) {
  var r = c.relative(te, t);
  if (!r.startsWith("..")) {
    t = r;
  }
  return text([ cmd(M.Style, L.Underline), t, cmd(M.Style, L.NoUnderline), "\n" ]);
}

function diagnosticMessage(t) {
  var r = L.Foreground;
  if ("info" === t.severity) {
    r = L.BrightBlue;
  } else if ("warn" === t.severity) {
    r = L.BrightYellow;
  } else if ("error" === t.severity) {
    r = L.BrightRed;
  }
  return text([ re, cmd(M.Style, L.BrightBlack), `${t.line}:${t.col}`, D.Tab, cmd(M.Style, r), t.severity, D.Tab, cmd(M.Style, L.Foreground), indent(t.message.trim(), text([ re, D.Tab, D.Tab ])), D.Newline ]);
}

function diagnosticMessageGithub(t) {
  githubAnnotation("warn" === t.severity ? "warning" : "error" === t.severity ? "error" : "notice", t.message, {
    file: t.file,
    line: t.line,
    col: t.col,
    endLine: t.endLine,
    endColumn: t.endColumn
  });
}

function runningDiagnostics(t, r) {
  var i = t ? r ? `(${t}/${r})` : `(${t})` : "";
  var frame = t => text([ cmd(M.Style, L.Magenta), q[t % q.length], " ", cmd(M.Style, L.Foreground), `Checking files${D.Ellipsis} `, cmd(M.Style, L.BrightBlack), i ]);
  return concat([ fromValue(frame(0)), pipe(interval(150), map((t => frame(t + 1)))) ]);
}

var isMinSeverity = (t, r) => {
  switch (t) {
   case "info":
    return "warn" !== r && "error" !== r;

   case "warn":
    return "error" !== r;

   case "error":
    return !0;
  }
};

async function* run$1(t, i) {
  var {runDiagnostics: n} = await Promise.resolve().then((function() {
    return require("./thread-chunk3.js");
  }));
  var a;
  var o;
  try {
    a = await r.loadConfig(i.tsconfig);
    o = r.parseConfig(a.pluginConfig, a.rootPath);
  } catch (t) {
    throw externalError("Failed to load configuration.", t);
  }
  var s = {
    warn: 0,
    error: 0,
    info: 0
  };
  var l = i.minSeverity;
  var c = n({
    rootPath: a.rootPath,
    configPath: a.configPath,
    pluginConfig: o
  });
  var u = 0;
  var d = 0;
  try {
    if (t.isInteractive) {
      yield runningDiagnostics();
    }
    for await (var f of c) {
      if ("EXTERNAL_WARNING" === f.kind) {
        yield experimentMessage(`${code(".vue")} and ${code(".svelte")} file support is experimental.`);
      } else if ("FILE_COUNT" === f.kind) {
        u = f.fileCount;
      } else {
        d++;
        var g = "";
        for (var p of f.messages) {
          s[p.severity]++;
          if (isMinSeverity(p.severity, l)) {
            g += diagnosticMessage(p);
            diagnosticMessageGithub(p);
          }
        }
        if (g) {
          yield diagnosticFile(f.filePath) + g + "\n";
        }
      }
      if (t.isInteractive) {
        yield runningDiagnostics(d, u);
      }
    }
  } catch (t) {
    throw externalError("Could not check files", t);
  }
  if ("info" !== l) {
    s.info = 0;
  }
  if (i.failOnWarn && s.warn || s.error) {
    throw function problemsSummary(t) {
      var {info: r, error: i, warn: n} = t;
      var a = "";
      if (r) {
        a += text([ cmd(M.Style, L.Blue), U.Info, ` ${r} notices\n` ]);
      }
      return error(a += text([ cmd(M.Style, L.Red), U.Cross, ` ${i + n} problems (${i} errors, ${n} warnings)\n` ]));
    }(s);
  } else {
    yield function infoSummary(t) {
      var {info: r, error: i, warn: n} = t;
      var a = "";
      if (r) {
        a += text([ cmd(M.Style, L.Blue), U.Info, ` ${r} notices\n` ]);
      }
      if (i || n) {
        a += text([ cmd(M.Style, L.BrightYellow), U.Warning, ` ${i + n} problems (${i} errors, ${n} warnings)\n` ]);
      } else {
        a += text([ cmd(M.Style, L.BrightGreen), U.Tick, " No problems found\n" ]);
      }
      return a;
    }(s);
  }
}

class CheckCommand extends Command {
  static paths=[ [ "check" ] ];
  tsconfig=String$1("--tsconfig,-c", {
    description: "Specify the `tsconfig.json` used to read"
  });
  failOnWarn=Boolean("--fail-on-warn,-w", !1, {
    description: "Triggers an error and a non-zero exit code if any warnings have been reported"
  });
  minSeverity=String$1("--level,-l", {
    description: "The minimum severity of diagnostics to display (info, warn, error)",
    validator: isOneOf([ isLiteral("info"), isLiteral("warn"), isLiteral("error") ])
  }) || "info";
  async execute() {
    var t = initTTY();
    var r = await t.start(run$1(t, {
      failOnWarn: this.failOnWarn,
      minSeverity: this.minSeverity,
      tsconfig: this.tsconfig
    }));
    return exitCode() || ("object" == typeof r ? r.exit : 0);
  }
}

var ie = new Intl.Collator(0, {
  numeric: 1
}).compare;

var ne = {
  typescript_embed_lsp: "5.5.0",
  typescript: "4.1.0",
  tada: "1.0.0",
  lsp: "1.0.0"
};

var semverComply = (t, r) => {
  var i = t.match(/\d+\.\d+\.\d+/);
  return i ? function semiver(t, r, i) {
    t = t.split(".");
    r = r.split(".");
    return ie(t[0], r[0]) || ie(t[1], r[1]) || (r[2] = r.slice(2).join("."), (i = /[.-]/.test(t[2] = t.slice(2).join("."))) == /[.-]/.test(r[2]) ? ie(t[2], r[2]) : i ? -1 : 1);
  }(i[0], r) >= 0 : !1;
};

var spanToFilePosition = (t, r) => {
  var i = {
    fileName: t.fileName,
    line: 1,
    col: 1,
    endLine: void 0,
    endColumn: void 0
  };
  var n = t.getLineAndCharacterOfPosition("number" == typeof r ? r : r.start);
  i.line = n.line + 1;
  i.col = n.character + 1;
  if ("number" != typeof r && r.length > 1) {
    n = t.getLineAndCharacterOfPosition(r.start + r.length - 1);
    i.endLine = n.line + 1;
    i.endColumn = n.character + 1;
  }
  return i;
};

/**!
 * The MIT License (MIT)
 * Copyright (c) Microsoft Corporation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
 * associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
 * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 * @see {@link https://github.com/microsoft/TypeScript-Website/blob/7f1c6e0577a0df9f945530a876ec87145a8b1fee/packages/typescript-vfs/src/index.ts}
 */ function notImplemented(t) {
  throw new Error(`Method '${t}' is not implemented.`);
}

var defaultCompilerOptions = t => ({
  ...t.getDefaultCompilerOptions(),
  jsx: t.JsxEmit.React,
  strict: !0,
  esModuleInterop: !0,
  module: t.ModuleKind.ESNext,
  suppressOutputPathCheck: !0,
  skipLibCheck: !0,
  skipDefaultLibCheck: !0,
  moduleResolution: t.ModuleResolutionKind.NodeJs
});

function createVirtualCompilerHost(t, r, i) {
  var n = new Map;
  return {
    compilerHost: {
      ...t,
      getCanonicalFileName: t => t,
      getDefaultLibFileName: () => "/" + i.getDefaultLibFileName(r),
      getDirectories: () => [],
      getNewLine: () => t.newLine,
      getSourceFile: a => n.get(a) || (t => {
        n.set(t.fileName, t);
        return t;
      })(i.createSourceFile(a, t.readFile(a), r.target || defaultCompilerOptions(i).target, !1)),
      useCaseSensitiveFileNames: () => t.useCaseSensitiveFileNames
    },
    updateFile: r => {
      var i = n.has(r.fileName);
      t.writeFile(r.fileName, r.text);
      n.set(r.fileName, r);
      return i;
    }
  };
}

function createVirtualLanguageServiceHost(t, r, i, n, a) {
  var o = [ ...r ];
  var {compilerHost: s, updateFile: l} = createVirtualCompilerHost(t, i, n);
  var c = new Map;
  var u = 0;
  return {
    languageServiceHost: {
      ...s,
      getProjectVersion: () => u.toString(),
      getCompilationSettings: () => i,
      getCustomTransformers: () => a,
      getScriptFileNames: () => o.slice(),
      getScriptSnapshot: r => {
        var i = t.readFile(r);
        if (i) {
          return n.ScriptSnapshot.fromString(i);
        }
        return;
      },
      getScriptVersion: t => c.get(t) || "0",
      writeFile: t.writeFile
    },
    updateFile: t => {
      u++;
      c.set(t.fileName, u.toString());
      if (!o.includes(t.fileName)) {
        o.push(t.fileName);
      }
      l(t);
    }
  };
}

var ae = {};

var oe = {};

var se = {};

Object.defineProperty(se, "__esModule", {
  value: !0
});

se.binarySearch = void 0;

se.binarySearch = function binarySearch$2(t, r) {
  var i = 0;
  var n = t.length - 1;
  while (i <= n) {
    var a = Math.floor((i + n) / 2);
    var o = t[a];
    if (o < r) {
      i = a + 1;
    } else if (o > r) {
      n = a - 1;
    } else {
      i = a;
      n = a;
      break;
    }
  }
  return Math.max(Math.min(i, n, t.length - 1), 0);
};

var le = {};

Object.defineProperty(le, "__esModule", {
  value: !0
});

le.getStack = le.track = le.resetOffsetStack = le.offsetStack = le.setTracking = void 0;

var ce = !0;

var ue = 0;

le.setTracking = function setTracking(t) {
  ce = t;
};

le.offsetStack = function offsetStack() {
  ue++;
};

le.resetOffsetStack = function resetOffsetStack() {
  ue--;
};

le.track = function track(t, r = []) {
  return [ new Proxy(t, {
    get(t, r, i) {
      if (ce) {
        if ("push" === r) {
          return push;
        }
        if ("pop" === r) {
          return pop;
        }
        if ("shift" === r) {
          return shift;
        }
        if ("unshift" === r) {
          return unshift;
        }
        if ("splice" === r) {
          return splice;
        }
        if ("sort" === r) {
          return sort;
        }
        if ("reverse" === r) {
          return reverse;
        }
      }
      return Reflect.get(t, r, i);
    }
  }), r ];
  function push(...i) {
    r.push({
      stack: getStack(),
      length: i.length
    });
    return t.push(...i);
  }
  function pop() {
    if (r.length) {
      var i = r[r.length - 1];
      if (i.length > 1) {
        i.length--;
      } else {
        r.pop();
      }
    }
    return t.pop();
  }
  function shift() {
    if (r.length) {
      var i = r[0];
      if (i.length > 1) {
        i.length--;
      } else {
        r.shift();
      }
    }
    return t.shift();
  }
  function unshift(...i) {
    r.unshift({
      stack: getStack(),
      length: i.length
    });
    return t.unshift(...i);
  }
  function splice(i, n, ...a) {
    if (void 0 === n) {
      n = t.length - i;
    }
    var o = 0;
    var s;
    for (var l = 0; l < r.length; l++) {
      var c = r[l];
      var u = o;
      o = u + c.length;
      if (i >= u) {
        s = l + 1;
        var d = c.length;
        c.length = i - u;
        r.splice(s, 0, {
          stack: c.stack,
          length: d - c.length
        });
        break;
      }
    }
    if (void 0 === s) {
      throw new Error("Invalid splice operation");
    }
    var f = n;
    for (var g = s; g < r.length; g++) {
      var p = r[g];
      while (f > 0 && p.length > 0) {
        p.length--;
        f--;
      }
      if (0 === f) {
        break;
      }
    }
    r.splice(s, 0, {
      stack: getStack(),
      length: a.length
    });
    return t.splice(i, n, ...a);
  }
  function sort(i) {
    r.splice(0, r.length, {
      stack: getStack(),
      length: t.length
    });
    return t.sort(i);
  }
  function reverse() {
    r.splice(0, r.length, {
      stack: getStack(),
      length: t.length
    });
    return t.reverse();
  }
};

function getStack() {
  var t = (new Error).stack.split("\n")[3 + ue].trim();
  if (t.endsWith(")")) {
    t = t.slice(t.lastIndexOf("(") + 1, -1);
  } else {
    t = t.slice(t.lastIndexOf(" ") + 1);
  }
  return t;
}

le.getStack = getStack;

var de = {};

Object.defineProperty(de, "__esModule", {
  value: !0
});

!function(t) {
  var r = this && this.__createBinding || (Object.create ? function(t, r, i, n) {
    if (void 0 === n) {
      n = i;
    }
    var a = Object.getOwnPropertyDescriptor(r, i);
    if (!a || ("get" in a ? !r.__esModule : a.writable || a.configurable)) {
      a = {
        enumerable: !0,
        get: function() {
          return r[i];
        }
      };
    }
    Object.defineProperty(t, n, a);
  } : function(t, r, i, n) {
    if (void 0 === n) {
      n = i;
    }
    t[n] = r[i];
  });
  var i = this && this.__exportStar || function(t, i) {
    for (var n in t) {
      if ("default" !== n && !Object.prototype.hasOwnProperty.call(i, n)) {
        r(i, t, n);
      }
    }
  };
  Object.defineProperty(t, "__esModule", {
    value: !0
  });
  t.replaceRange = t.replaceSourceRange = t.replaceAll = t.replace = t.create = t.toString = t.getLength = void 0;
  var n = se;
  var a = le;
  i(de, t);
  i(le, t);
  function getLength(t) {
    var r = 0;
    for (var i of t) {
      r += "string" == typeof i ? i.length : i[0].length;
    }
    return r;
  }
  t.getLength = getLength;
  function toString(t) {
    return t.map((t => "string" == typeof t ? t : t[0])).join("");
  }
  t.toString = toString;
  t.create = function create(t) {
    return [ [ t, void 0, 0 ] ];
  };
  t.replace = function replace(t, r, ...i) {
    var n = toString(t).match(r);
    if (n && void 0 !== n.index) {
      var o = n.index;
      var s = o + n[0].length;
      (0, a.offsetStack)();
      replaceRange(t, o, s, ...i.map((t => "function" == typeof t ? t(n[0]) : t)));
      (0, a.resetOffsetStack)();
    }
  };
  t.replaceAll = function replaceAll(t, r, ...i) {
    var n = toString(t);
    var o = n.matchAll(r);
    var s = n.length;
    var l = 0;
    var _loop = function(r) {
      if (void 0 !== r.index) {
        var n = r.index + l;
        var o = n + r[0].length;
        (0, a.offsetStack)();
        replaceRange(t, n, o, ...i.map((t => "function" == typeof t ? t(r[0]) : t)));
        (0, a.resetOffsetStack)();
        var c = getLength(t);
        l += c - s;
        s = c;
      }
    };
    for (var c of o) {
      _loop(c);
    }
  };
  t.replaceSourceRange = function replaceSourceRange(t, r, i, n, ...o) {
    for (var s of t) {
      if ("string" == typeof s) {
        continue;
      }
      if (s[1] === r) {
        var l = s[2];
        var c = s[2] + s[0].length;
        if (l <= i && c >= n) {
          var u = [];
          if (i > l) {
            u.push(trimSegmentEnd(s, i - l));
          }
          for (var d of o) {
            u.push(d);
          }
          if (n < c) {
            u.push(trimSegmentStart(s, n - c));
          }
          combineStrings(u);
          (0, a.offsetStack)();
          t.splice(t.indexOf(s), 1, ...u);
          (0, a.resetOffsetStack)();
          return !0;
        }
      }
    }
    return !1;
  };
  function replaceRange(t, r, i, ...o) {
    var s = function toOffsets(t) {
      var r = [];
      var i = 0;
      for (var n of t) {
        r.push(i);
        i += "string" == typeof n ? n.length : n[0].length;
      }
      return r;
    }(t);
    var l = (0, n.binarySearch)(s, r);
    var c = (0, n.binarySearch)(s, i);
    var u = t[c];
    var d = s[l];
    var f = s[c];
    var g = s[c] + ("string" == typeof u ? u.length : u[0].length);
    var p = [];
    if (r > d) {
      p.push(trimSegmentEnd(t[l], r - d));
    }
    for (var h of o) {
      p.push(h);
    }
    if (i < g) {
      p.push(trimSegmentStart(u, i - f));
    }
    combineStrings(p);
    (0, a.offsetStack)();
    t.splice(l, c - l + 1, ...p);
    (0, a.resetOffsetStack)();
  }
  t.replaceRange = replaceRange;
  function combineStrings(t) {
    for (var r = t.length - 1; r >= 1; r--) {
      if ("string" == typeof t[r] && "string" == typeof t[r - 1]) {
        t[r - 1] = t[r - 1] + t[r];
        (0, a.offsetStack)();
        t.splice(r, 1);
        (0, a.resetOffsetStack)();
      }
    }
  }
  function trimSegmentEnd(t, r) {
    if ("string" == typeof t) {
      return t.slice(0, r);
    }
    return [ t[0].slice(0, r), ...t.slice(1) ];
  }
  function trimSegmentStart(t, r) {
    if ("string" == typeof t) {
      return t.slice(r);
    }
    if (r < 0) {
      r += t[0].length;
    }
    return [ t[0].slice(r), t[1], t[2] + r, ...t.slice(3) ];
  }
}(oe);

var fe = {};

var ge = {};

Object.defineProperty(ge, "__esModule", {
  value: !0
});

ge.binarySearch = void 0;

ge.binarySearch = function binarySearch(t, r) {
  var i = 0;
  var n = t.length - 1;
  var a;
  while (i <= n) {
    var o = Math.floor((i + n) / 2);
    var s = t[o];
    if (s < r) {
      i = o + 1;
    } else if (s > r) {
      n = o - 1;
    } else {
      i = o;
      n = o;
      a = o;
      break;
    }
  }
  return {
    low: Math.max(Math.min(i, n, t.length - 1), 0),
    high: Math.min(Math.max(i, n, 0), t.length - 1),
    match: a
  };
};

var pe = {};

Object.defineProperty(pe, "__esModule", {
  value: !0
});

pe.translateOffset = void 0;

pe.translateOffset = function translateOffset(t, r, i, n) {
  for (var a = 0; a < r.length; a++) {
    var o = r[a];
    if (t >= o && t <= o + n[a]) {
      return i[a] + t - o;
    }
  }
};

Object.defineProperty(fe, "__esModule", {
  value: !0
});

fe.SourceMap = void 0;

var he = ge;

var ve = pe;

fe.SourceMap = class SourceMap {
  constructor(t) {
    this.mappings = t;
  }
  getSourceOffset(t) {
    for (var r of this.findMatching(t, "generatedOffsets", "sourceOffsets")) {
      return r;
    }
  }
  getGeneratedOffset(t) {
    for (var r of this.findMatching(t, "sourceOffsets", "generatedOffsets")) {
      return r;
    }
  }
  getSourceOffsets(t) {
    return this.findMatching(t, "generatedOffsets", "sourceOffsets");
  }
  getGeneratedOffsets(t) {
    return this.findMatching(t, "sourceOffsets", "generatedOffsets");
  }
  * findMatching(t, r, i) {
    var n = this.getMemoBasedOnRange(r);
    if (0 === n.offsets.length) {
      return;
    }
    var {low: a, high: o} = (0, he.binarySearch)(n.offsets, t);
    var s = new Set;
    for (var l = a; l <= o; l++) {
      for (var c of n.mappings[l]) {
        if (s.has(c)) {
          continue;
        }
        s.add(c);
        var u = (0, ve.translateOffset)(t, c[r], c[i], c.lengths);
        if (void 0 !== u) {
          yield [ u, c ];
        }
      }
    }
  }
  getMemoBasedOnRange(t) {
    return "sourceOffsets" === t ? this.sourceCodeOffsetsMemo ??= this.createMemo("sourceOffsets") : this.generatedCodeOffsetsMemo ??= this.createMemo("generatedOffsets");
  }
  createMemo(t) {
    var r = new Set;
    for (var i of this.mappings) {
      for (var n = 0; n < i[t].length; n++) {
        r.add(i[t][n]);
        r.add(i[t][n] + i.lengths[n]);
      }
    }
    var a = [ ...r ].sort(((t, r) => t - r));
    var o = a.map((() => new Set));
    for (var s of this.mappings) {
      for (var l = 0; l < s[t].length; l++) {
        var c = (0, he.binarySearch)(a, s[t][l]).match;
        var u = (0, he.binarySearch)(a, s[t][l] + s.lengths[l]).match;
        for (var d = c; d <= u; d++) {
          o[d].add(s);
        }
      }
    }
    return {
      offsets: a,
      mappings: o
    };
  }
};

var me = {};

Object.defineProperty(me, "__esModule", {
  value: !0
});

me.buildMappings = void 0;

me.buildMappings = function buildMappings(t) {
  var r = 0;
  var i = [];
  for (var n of t) {
    if ("string" == typeof n) {
      r += n.length;
    } else {
      i.push({
        source: n[1],
        sourceOffsets: [ n[2] ],
        generatedOffsets: [ r ],
        lengths: [ n[0].length ],
        data: n[3]
      });
      r += n[0].length;
    }
  }
  return i;
};

var ye = {};

Object.defineProperty(ye, "__esModule", {
  value: !0
});

ye.buildStacks = void 0;

ye.buildStacks = function buildStacks(t, r) {
  var i = 0;
  var n = 0;
  var a = [];
  for (var o of r) {
    var s = i;
    for (var l = 0; l < o.length; l++) {
      var c = t[n + l];
      if ("string" == typeof c) {
        i += c.length;
      } else {
        i += c[0].length;
      }
    }
    n += o.length;
    a.push({
      source: o.stack,
      range: [ s, i ]
    });
  }
  return a;
};

!function(t) {
  var r = this && this.__createBinding || (Object.create ? function(t, r, i, n) {
    if (void 0 === n) {
      n = i;
    }
    var a = Object.getOwnPropertyDescriptor(r, i);
    if (!a || ("get" in a ? !r.__esModule : a.writable || a.configurable)) {
      a = {
        enumerable: !0,
        get: function() {
          return r[i];
        }
      };
    }
    Object.defineProperty(t, n, a);
  } : function(t, r, i, n) {
    if (void 0 === n) {
      n = i;
    }
    t[n] = r[i];
  });
  var i = this && this.__exportStar || function(t, i) {
    for (var n in t) {
      if ("default" !== n && !Object.prototype.hasOwnProperty.call(i, n)) {
        r(i, t, n);
      }
    }
  };
  Object.defineProperty(t, "__esModule", {
    value: !0
  });
  i(oe, t);
  i(fe, t);
  i(pe, t);
  i(me, t);
  i(ye, t);
}(ae);

class SourceMappedFile extends ae.SourceMap {
  constructor(t, r) {
    super(t);
    this.sourceFile = r.sourceFile;
    this.sourceFileId = r.sourceFileId;
    this.generatedFileId = r.generatedFileId;
  }
  getSourceSpan(t) {
    var r = this.getSourceOffset("number" == typeof t ? t : t.start);
    if (null != r) {
      var i = "number" != typeof t ? this.getSourceOffset(t.start + t.length - 1) : null;
      return {
        fileName: this.sourceFileId,
        start: r[0],
        length: null != i ? i[0] - r[0] + 1 : "number" != typeof t ? t.length : 1
      };
    } else {
      return {
        fileName: this.generatedFileId,
        start: "number" != typeof t ? t.start : t,
        length: "number" != typeof t ? t.length : 1
      };
    }
  }
}

function maybeBind(t, r) {
  return r ? r.bind(t) : void 0;
}

var buildLanguageService = t => {
  var {virtualMap: r} = t;
  var i = s.createLanguageService(t.languageServiceHost, t.documentRegistry);
  var n = maybeBind(i, i.getProgram);
  var mapFileFn = t => (n, ...a) => {
    var o = r.get(n);
    if (o && o.sourceFileId === n) {
      n = o.generatedFileId;
    }
    return t.call(i, n, ...a);
  };
  var mapFileTextSpanFn = t => (n, a, ...o) => {
    var s = r.get(n);
    if (s && s.sourceFileId === n) {
      n = s.generatedFileId;
      var l = s.getGeneratedOffset(a.start)?.[0];
      var c = s.getGeneratedOffset(a.start + a.length - 1)?.[0];
      if (null != l) {
        a = {
          start: l,
          length: a.length
        };
        if (null != c) {
          a.length = c - l + 1;
        }
      }
    }
    return t.call(i, n, a, ...o);
  };
  var mapFileTextRangeFn = t => (n, a, ...o) => {
    var s = r.get(n);
    if (s && s.sourceFileId === n) {
      n = s.generatedFileId;
      var l = s.getGeneratedOffset(a.pos)?.[0];
      var c = s.getGeneratedOffset(a.end)?.[0];
      if (null != l) {
        a = {
          pos: l,
          end: l + (a.end - a.pos)
        };
        if (null != c) {
          a.end = c;
        }
      }
    }
    return t.call(i, n, a, ...o);
  };
  var mapFilePositionFn = t => (n, a, ...o) => {
    var s = r.get(n);
    if (s && s.sourceFileId === n) {
      n = s.generatedFileId;
      a = s.getGeneratedOffset(a)?.[0] ?? a;
    }
    return t.call(i, n, a, ...o);
  };
  var mapFileStartEndFn = t => (n, a, o, ...s) => {
    var l = r.get(n);
    if (l && l.sourceFileId === n) {
      n = l.generatedFileId;
      a = l.getGeneratedOffset(a)?.[0] ?? a;
      o = l.getGeneratedOffset(o)?.[0] ?? o;
    }
    return t.call(i, n, a, o, ...s);
  };
  var a;
  return Object.assign(i, {
    getProgram() {
      if (a) {
        return a;
      } else {
        var r = n();
        return r ? a = (t => {
          var {program: r, virtualMap: i, projectRoot: n} = t;
          var a = maybeBind(r, r.isSourceFileFromExternalLibrary);
          var o = maybeBind(r, r.getModeForResolutionAtIndex);
          var s = maybeBind(r, r.getSourceFile);
          var l = maybeBind(r, r.getSourceFiles);
          var mapSourceFileFn = t => (n, ...a) => {
            var o = n && i.get(n.fileName);
            if (o && o.sourceFileId === n?.fileName) {
              n = s(o.generatedFileId) || n;
            }
            return t.call(r, n, ...a);
          };
          return Object.assign(r, {
            getSyntacticDiagnostics: mapSourceFileFn(r.getSyntacticDiagnostics),
            getSemanticDiagnostics: mapSourceFileFn(r.getSemanticDiagnostics),
            getDeclarationDiagnostics: mapSourceFileFn(r.getDeclarationDiagnostics),
            isSourceFileDefaultLibrary: mapSourceFileFn(r.isSourceFileDefaultLibrary),
            getModeForUsageLocation: mapSourceFileFn(r.getModeForUsageLocation),
            isSourceFileFromExternalLibrary(t) {
              var r = i.get(t.fileName);
              return !(!r || r.sourceFileId !== t.fileName) || a(t);
            },
            getSourceFiles() {
              var t = [];
              for (var i of l()) {
                if (!c.relative(n, i.fileName).startsWith("..") && !r.isSourceFileFromExternalLibrary(i)) {
                  t.push(i);
                }
              }
              return t;
            },
            getSourceFile(t) {
              var r = i.get(t);
              if (r && r.sourceFileId === t) {
                t = r.generatedFileId;
              }
              return s(t);
            },
            getModeForResolutionAtIndex(t, r) {
              var n = i.get(t.fileName);
              if (n && n.sourceFileId === t.fileName) {
                t = s(n.generatedFileId) || t;
                r = n.getGeneratedOffset(r)?.[0] ?? r;
              }
              return o(t, r);
            }
          });
        })({
          program: r,
          virtualMap: t.virtualMap,
          projectRoot: t.projectRoot
        }) : void 0;
      }
    },
    getReferencesAtPosition: mapFilePositionFn(i.getReferencesAtPosition),
    findReferences: mapFilePositionFn(i.findReferences),
    getDefinitionAtPosition: mapFilePositionFn(i.getDefinitionAtPosition),
    getDefinitionAndBoundSpan: mapFilePositionFn(i.getDefinitionAndBoundSpan),
    getTypeDefinitionAtPosition: mapFilePositionFn(i.getTypeDefinitionAtPosition),
    getImplementationAtPosition: mapFilePositionFn(i.getImplementationAtPosition),
    getCompletionsAtPosition: mapFilePositionFn(i.getCompletionsAtPosition),
    getCompletionEntryDetails: mapFilePositionFn(i.getCompletionEntryDetails),
    getCompletionEntrySymbol: mapFilePositionFn(i.getCompletionEntrySymbol),
    getQuickInfoAtPosition: mapFilePositionFn(i.getQuickInfoAtPosition),
    getBreakpointStatementAtPosition: mapFilePositionFn(i.getBreakpointStatementAtPosition),
    getSignatureHelpItems: mapFilePositionFn(i.getSignatureHelpItems),
    getRenameInfo: mapFilePositionFn(i.getRenameInfo),
    getSmartSelectionRange: mapFilePositionFn(i.getSmartSelectionRange),
    getDocumentHighlights: mapFilePositionFn(i.getDocumentHighlights),
    prepareCallHierarchy: mapFilePositionFn(i.prepareCallHierarchy),
    provideCallHierarchyIncomingCalls: mapFilePositionFn(i.provideCallHierarchyIncomingCalls),
    provideCallHierarchyOutgoingCalls: mapFilePositionFn(i.provideCallHierarchyOutgoingCalls),
    getBraceMatchingAtPosition: mapFilePositionFn(i.getBraceMatchingAtPosition),
    getIndentationAtPosition: mapFilePositionFn(i.getIndentationAtPosition),
    getFormattingEditsAfterKeystroke: mapFilePositionFn(i.getFormattingEditsAfterKeystroke),
    getDocCommentTemplateAtPosition: mapFilePositionFn(i.getDocCommentTemplateAtPosition),
    isValidBraceCompletionAtPosition: mapFilePositionFn(i.isValidBraceCompletionAtPosition),
    getJsxClosingTagAtPosition: mapFilePositionFn(i.getJsxClosingTagAtPosition),
    getLinkedEditingRangeAtPosition: mapFilePositionFn(i.getLinkedEditingRangeAtPosition),
    getSpanOfEnclosingComment: mapFilePositionFn(i.getSpanOfEnclosingComment),
    getSyntacticClassifications: mapFileTextSpanFn(i.getSyntacticClassifications),
    getSemanticClassifications: mapFileTextSpanFn(i.getSemanticClassifications),
    getEncodedSyntacticClassifications: mapFileTextSpanFn(i.getEncodedSyntacticClassifications),
    getEncodedSemanticClassifications: mapFileTextSpanFn(i.getEncodedSemanticClassifications),
    provideInlayHints: mapFileTextSpanFn(i.provideInlayHints),
    getNameOrDottedNameSpan: mapFileStartEndFn(i.getNameOrDottedNameSpan),
    getFormattingEditsForRange: mapFileStartEndFn(i.getFormattingEditsForRange),
    getCodeFixesAtPosition: mapFileStartEndFn(i.getCodeFixesAtPosition),
    getFileReferences: mapFileFn(i.getFileReferences),
    getNavigationBarItems: mapFileFn(i.getNavigationBarItems),
    getNavigationTree: mapFileFn(i.getNavigationTree),
    getOutliningSpans: mapFileFn(i.getOutliningSpans),
    getTodoComments: mapFileFn(i.getTodoComments),
    getFormattingEditsForDocument: mapFileFn(i.getFormattingEditsForDocument),
    getEditsForRefactor: mapFileFn(i.getEditsForRefactor),
    getEmitOutput: mapFileFn(i.getEmitOutput),
    getSuggestionDiagnostics: mapFileFn(i.getSuggestionDiagnostics),
    getSemanticDiagnostics: mapFileFn(i.getSemanticDiagnostics),
    getSyntacticDiagnostics: mapFileFn(i.getSyntacticDiagnostics),
    getSupportedCodeFixes: mapFileFn(i.getSupportedCodeFixes),
    toggleLineComment: mapFileTextRangeFn(i.toggleLineComment),
    toggleMultilineComment: mapFileTextRangeFn(i.toggleMultilineComment),
    commentSelection: mapFileTextRangeFn(i.commentSelection),
    uncommentSelection: mapFileTextRangeFn(i.uncommentSelection)
  });
};

var buildPluginInfo = t => {
  var r;
  return {
    config: t.config,
    get languageService() {
      return t.getLanguageService();
    },
    get languageServiceHost() {
      return r || (r = createVirtualLanguageServiceHost(t.system, [ ...t.rootNames ], t.options, s).languageServiceHost);
    },
    project: {
      getProjectName: () => t.projectRoot,
      projectService: {
        logger: console
      }
    },
    serverHost: {
      ...t.system,
      setImmediate,
      clearImmediate
    }
  };
};

var Se;

var we;

var be = [ ".svelte", ".vue" ];

var transform = async t => {
  var r = c.extname(t.fileName);
  if (".svelte" === r) {
    return (async (...t) => {
      if (!Se) {
        try {
          Se = await import("@gql.tada/svelte-support");
        } catch (t) {
          throw new TadaError(I.SVELTE_SUPPORT, "For Svelte support the `@gql.tada/svelte-support` package must be installed.\nInstall the package and try again.");
        }
      }
      return Se.transform(...t);
    })(t);
  } else if (".vue" === r) {
    await (async () => {
      if (!we) {
        try {
          we = await import("@gql.tada/vue-support");
        } catch (t) {
          throw new TadaError(I.VUE_SUPPORT, "For Vue support the `@gql.tada/vue-support` package must be installed.\nInstall the package and try again.");
        }
      }
      return we.check();
    })();
    return (async (...t) => {
      if (!we) {
        try {
          we = await import("@gql.tada/vue-support");
        } catch (t) {
          throw new TadaError(I.VUE_SUPPORT, "For Vue support the `@gql.tada/vue-support` package must be installed.\nInstall the package and try again.");
        }
      }
      return we.transform(...t);
    })(t);
  } else {
    throw new TadaError(I.UNKNOWN_EXTERNAL_FILE, `Tried transforming unknown file type "${r}". Supported: ${be.join(", ")}`);
  }
};

var programFactory = t => {
  var r = new Map;
  var i = new Map;
  var n = function createFSBackedSystem(t, r, i, n) {
    var a = r + "/vfs";
    var o = i.sys;
    var s = n ?? c.dirname(require.resolve("typescript"));
    return {
      name: "fs-vfs",
      root: a,
      args: [],
      realpath: t => {
        if (o.realpath) {
          return o.realpath(t);
        }
        return t;
      },
      createDirectory: () => notImplemented("createDirectory"),
      directoryExists: r => Array.from(t.keys()).some((t => t.startsWith(r))) || o.directoryExists(r),
      exit: o.exit,
      fileExists: r => {
        if (t.has(r)) {
          return !0;
        }
        if (r.includes("tsconfig.json") || r.includes("tsconfig.json")) {
          return !1;
        }
        if (r.startsWith("/lib")) {
          var i = `${s}/${r.replace("/", "")}`;
          return o.fileExists(i);
        }
        return o.fileExists(r);
      },
      getCurrentDirectory: () => a,
      getDirectories: o.getDirectories,
      getExecutingFilePath: () => notImplemented("getExecutingFilePath"),
      readDirectory: (...r) => {
        if ("/" === r[0]) {
          return Array.from(t.keys());
        } else {
          return o.readDirectory(...r);
        }
      },
      readFile: r => {
        if (t.has(r)) {
          return t.get(r);
        }
        if (r.startsWith("/lib")) {
          var i = `${s}/${r.replace("/", "")}`;
          var n = o.readFile(i);
          if (!n) {
            var a = o.readDirectory(s);
            throw new Error(`TSVFS: A request was made for ${i} but there wasn't a file found in the file map. You likely have a mismatch in the compiler options for the CDN download vs the compiler program. Existing Libs: ${a}.`);
          }
          return n;
        }
        return o.readFile(r);
      },
      resolvePath: r => {
        if (t.has(r)) {
          return r;
        }
        return o.resolvePath(r);
      },
      newLine: "\n",
      useCaseSensitiveFileNames: !0,
      write: () => notImplemented("write"),
      writeFile: (r, i) => {
        t.set(r, i);
      }
    };
  }(r, t.rootPath, s, resolveDefaultLibsPath(t));
  var a = resolveConfig(t);
  var o = new Set(a.fileNames);
  var u = {
    ...s.getDefaultCompilerOptions(),
    getDefaultLibFilePath: s.getDefaultLibFilePath(a.options),
    noCheck: !0,
    ...a.options
  };
  var d = !1;
  if ("Bundler" in s.ModuleResolutionKind && (u.moduleResolution === s.ModuleResolutionKind.NodeNext || u.moduleResolution === s.ModuleResolutionKind.Node16)) {
    d = !0;
    u.moduleResolution = s.ModuleResolutionKind.Bundler;
  }
  var f = createVirtualCompilerHost(n, u, s);
  var g = s.createDocumentRegistry();
  var p = {
    get projectPath() {
      return t.rootPath;
    },
    get wasOriginallyNodeNext() {
      return d;
    },
    get projectDirectories() {
      var r = new Set([ t.rootPath ]);
      for (var i of o) {
        r.add(c.dirname(i));
      }
      return [ ...r ];
    },
    createSourceFile: (t, r) => s.createSourceFile(t.fileId, "object" == typeof t.sourceText ? t.sourceText.getText(0, t.sourceText.getLength()) : t.sourceText, u.target || s.ScriptTarget.ESNext, !0, r || (null != t.scriptKind ? t.scriptKind : s.ScriptKind.TSX)),
    createExternalFiles(r = be) {
      var i = [];
      var a = new Set(o);
      var l = new Set([ t.rootPath ]);
      for (var u of o) {
        l.add(c.dirname(u));
      }
      for (var d of l) {
        for (var f of n.readDirectory(d, r, [ "**/node_modules" ])) {
          if (!a.has(f)) {
            a.add(f);
            var g = n.readFile(f, "utf8");
            if (g) {
              i.push(p.createSourceFile({
                fileId: f,
                sourceText: g
              }, s.ScriptKind.External));
            }
          }
        }
      }
      return i;
    },
    addSourceFile(t) {
      var r = "fileName" in t ? t : p.createSourceFile(t, s.ScriptKind.TSX);
      f.updateFile(r);
      o.add(r.fileName);
      return p;
    },
    addMappedFile(t, r) {
      var n = "fileName" in t ? t : p.createSourceFile(t, s.ScriptKind.External);
      if (r.mappings.length) {
        o.delete(n.fileName);
      }
      var a = new SourceMappedFile(r.mappings, {
        sourceFile: n,
        sourceFileId: n.fileName,
        generatedFileId: r.fileId
      });
      i.set(a.sourceFileId, a);
      i.set(a.generatedFileId, a);
      return p;
    },
    async addVirtualFiles(t) {
      for (var r of t) {
        var i = `${r.fileName}.ts`;
        var n = await transform(r);
        if (n) {
          p.addSourceFile({
            fileId: i,
            sourceText: n.snapshot
          }).addMappedFile(r, {
            mappings: n.mappings,
            fileId: i
          });
        }
      }
      return p;
    },
    resolveModuleName(t, r, i) {
      if (!t.startsWith(".")) {
        return t;
      }
      var n = s.resolveModuleName(t, r, u, f.compilerHost);
      if (n.resolvedModule) {
        var a = n.resolvedModule.resolvedFileName;
        var o = c.dirname(i);
        var l = c.relative(o, a);
        if (!l.startsWith(".")) {
          l = "./" + l;
        }
        return l.replace(/\\/g, "/");
      }
      var d = c.dirname(r);
      var g = c.dirname(i);
      var p = c.resolve(d, t);
      var h = c.relative(g, p);
      if (!h.startsWith(".")) {
        h = "./" + h;
      }
      return h.replace(/\\/g, "/");
    },
    build() {
      (() => {
        if ("stackTraceLimit" in Error && Error.stackTraceLimit < 25) {
          Error.stackTraceLimit = 25;
        }
      })();
      l.init({
        typescript: s
      });
      return (t => {
        var r;
        var i;
        var n;
        var a;
        var getLanguageServiceHost = () => n || (n = createVirtualLanguageServiceHost(t.system, [ ...t.rootNames ], t.options, s).languageServiceHost);
        var getLanguageService = () => i || (i = buildLanguageService({
          system: t.system,
          rootNames: t.rootNames,
          virtualMap: t.virtualMap,
          options: t.options,
          projectRoot: t.projectRoot,
          languageServiceHost: getLanguageServiceHost(),
          documentRegistry: t.documentRegistry
        }));
        var getProgram = () => {
          if (!r) {
            r = getLanguageService().getProgram() || s.createProgram({
              rootNames: t.rootNames,
              options: t.options,
              host: t.compilerHost
            });
          }
          return r;
        };
        return {
          get program() {
            return getProgram();
          },
          get languageService() {
            return getLanguageService();
          },
          buildPluginInfo: r => a || (a = buildPluginInfo({
            getLanguageService,
            getLanguageServiceHost,
            projectRoot: t.projectRoot,
            rootNames: t.rootNames,
            system: t.system,
            options: t.options,
            config: r
          })),
          getSourceFile: t => getProgram().getSourceFile(t),
          getSourceFiles: () => getProgram().getSourceFiles(),
          getSourceMappedFile: r => t.virtualMap.get("string" != typeof r ? r.fileName : r),
          getSourceSpan(r, i) {
            var n = "string" != typeof r ? r.fileName : r;
            var a = t.virtualMap.get(n);
            return a ? a.getSourceSpan(i) : {
              fileName: n,
              start: "number" != typeof i ? i.start : i,
              length: "number" != typeof i ? i.length : 1
            };
          },
          getSourcePosition(r, i) {
            var n = t.virtualMap.get("string" != typeof r ? r.fileName : r);
            if (n) {
              var a = n.getSourceSpan(i);
              if (a.fileName === n.sourceFileId) {
                return spanToFilePosition(n.sourceFile, a);
              }
            }
            return spanToFilePosition("string" == typeof r ? getProgram().getSourceFile(r) : r, i);
          }
        };
      })({
        virtualMap: i,
        projectRoot: t.rootPath,
        compilerHost: f.compilerHost,
        rootNames: [ ...o ],
        options: u,
        system: n,
        documentRegistry: g
      });
    }
  };
  return p;
};

var resolveDefaultLibsPath = t => {
  var r = s.getDefaultLibFilePath({});
  if (!s.sys.fileExists(r)) {
    var i = o.createRequire(t.configPath);
    var n = "typescript/package.json";
    try {
      return c.dirname(i.resolve(n, {
        paths: [ c.join(c.dirname(t.configPath), "node_modules"), c.join(t.rootPath, "node_modules"), ...i.resolve.paths(n) || [] ]
      }));
    } catch (r) {
      return c.resolve(t.rootPath, "node_modules", "typescript", "lib");
    }
  } else {
    return c.dirname(r);
  }
};

var resolveConfig = t => {
  var r = s.sys.readFile(t.configPath, "utf8") || "{}";
  var i = s.parseConfigFileTextToJson(t.configPath, r);
  if (null != i.error) {
    throw new Error(i.error.messageText.toString());
  }
  var n = c.dirname(t.configPath);
  return s.parseJsonConfigFileContent(i.config, s.sys, n, s.getDefaultCompilerOptions(), t.configPath);
};

var xe = function(t) {
  t[t.File = 0] = "File";
  t[t.Directory = 1] = "Directory";
  return t;
}({});

var stat = async (t, r = xe.File) => await u.stat(t).then((t => {
  switch (r) {
   case xe.File:
    return t.isFile();

   case xe.Directory:
    return t.isDirectory();
  }
})).catch((() => !1));

var Ee = /^(?:graphql\.config|\.graphqlrc)\.(?:cjs|[jt]s|json|toml|ya?ml)$/i;

var findGraphQLConfig = async t => {
  var r = process.cwd();
  var i = c.resolve(r, "/");
  while (r !== i) {
    var n = [];
    try {
      n = await u.readdir(r);
    } catch (t) {}
    var a = n.find((t => Ee.test(t)));
    if (a) {
      return a;
    }
    if (await stat(c.resolve(r, ".git"), xe.Directory)) {
      break;
    } else if (await stat(c.resolve(r, ".vscode"), xe.Directory)) {
      break;
    }
    r = c.resolve(r, "..");
  }
  var o = await (async t => {
    var r = t || process.cwd();
    var i = c.resolve(r, "/");
    while (r !== i) {
      if (await stat(c.resolve(r, ".git"), xe.Directory)) {
        return r;
      } else if (await stat(c.resolve(r, ".vscode"), xe.Directory)) {
        return r;
      } else if (await stat(c.resolve(r, "pnpm-workspace.yml"))) {
        return r;
      }
      var n = c.resolve(r, "package.json");
      if (await stat(n)) {
        try {
          var a = JSON.parse(await u.readFile(n, "utf8"));
          if (a && "object" == typeof a && Array.isArray(a.workspaces)) {
            return r;
          }
        } catch (t) {}
      }
      r = c.resolve(r, "..");
    }
    return null;
  })(t);
  if (o) {
    var s = c.resolve(o, "package.json");
    if (await stat(s)) {
      try {
        var l = JSON.parse(await u.readFile(s, "utf8"));
        if (l && "object" == typeof l && "graphql" in l && l.graphql && "object" == typeof l.graphql) {
          return s;
        }
      } catch (t) {}
    }
  }
  return null;
};

var loadPnpmWorkspaceCatalogs = async () => {
  var r = await (async (r = process.cwd()) => {
    var n = r;
    var a = t.parse(n).root;
    while (n !== a) {
      var o = t.join(n, "pnpm-workspace.yaml");
      try {
        await i.access(o);
        return n;
      } catch {}
      n = t.dirname(n);
    }
    return null;
  })();
  if (!r) {
    return null;
  }
  try {
    var n = t.join(r, "pnpm-workspace.yaml");
    return (t => {
      var r = t.split("\n");
      var i = {};
      var n = null;
      var a = null;
      for (var o of r) {
        var s = o.trim();
        if (!s || s.startsWith("#")) {
          continue;
        }
        if (s.endsWith(":") && !s.includes(" ")) {
          var l = s.slice(0, -1).replace(/['"]/g, "");
          if ("packages" === l) {
            n = "packages";
            i.packages = [];
          } else if ("catalog" === l) {
            n = "catalog";
            i.catalog = {};
          } else if ("catalogs" === l) {
            n = "catalogs";
            i.catalogs = {};
          } else if ("catalogs" === n) {
            a = l;
            i.catalogs[l] = {};
          }
        } else if (s.startsWith("- ")) {
          if ("packages" === n) {
            i.packages.push(s.slice(2).replace(/['"]/g, ""));
          }
        } else if (s.includes(":")) {
          var c = s.indexOf(":");
          var u = s.slice(0, c).trim().replace(/['"]/g, "");
          var d = s.slice(c + 1).trim().replace(/['"]/g, "");
          if ("catalog" === n) {
            i.catalog[u] = d;
          } else if ("catalogs" === n && a) {
            i.catalogs[a][u] = d;
          }
        }
      }
      return i;
    })(await i.readFile(n, "utf-8"));
  } catch {
    return null;
  }
};

var loadBunWorkspaceCatalogs = async () => {
  var r = await (async (r = process.cwd()) => {
    var n = r;
    var a = t.parse(n).root;
    while (n !== a) {
      var o = t.join(n, "package.json");
      try {
        await i.access(o);
        var s = await i.readFile(o, "utf-8");
        var l = JSON.parse(s);
        if (l.workspaces && (l.workspaces.catalog || l.workspaces.catalogs)) {
          return n;
        }
      } catch {}
      n = t.dirname(n);
    }
    return null;
  })();
  if (!r) {
    return null;
  }
  try {
    var n = t.join(r, "package.json");
    var a = await i.readFile(n, "utf-8");
    var o = JSON.parse(a);
    if (!o.workspaces) {
      return null;
    }
    var s = {};
    if (o.workspaces.packages) {
      s.packages = o.workspaces.packages;
    }
    if (o.workspaces.catalog) {
      s.catalog = o.workspaces.catalog;
    }
    if (o.workspaces.catalogs) {
      s.catalogs = o.workspaces.catalogs;
    }
    return s;
  } catch {
    return null;
  }
};

var loadWorkspaceCatalogs = async () => {
  var t = await loadPnpmWorkspaceCatalogs();
  if (t) {
    return t;
  }
  var r = await loadBunWorkspaceCatalogs();
  if (r) {
    return r;
  }
  return null;
};

var resolveCatalogVersion = t => {
  if (!t.startsWith("catalog:")) {
    return null;
  }
  var r = t.slice(8);
  if (!r || "default" === r) {
    return null;
  }
  return r;
};

var getVersionFromCatalog = (t, r, i) => {
  if (r) {
    return i.catalogs?.[r]?.[t] || null;
  } else {
    return i.catalog?.[t] || null;
  }
};

var getPackageVersion = async (t, r) => {
  var i = r.devDependencies?.[t];
  if (i) {
    if (i.startsWith("catalog:")) {
      var n = await loadWorkspaceCatalogs();
      if (n) {
        var a = resolveCatalogVersion(i);
        var o = getVersionFromCatalog(t, a, n);
        if (o) {
          return o;
        }
      }
    } else {
      return i;
    }
  }
  var s = r.dependencies?.[t];
  if (s) {
    if (s.startsWith("catalog:")) {
      var l = await loadWorkspaceCatalogs();
      if (l) {
        var c = resolveCatalogVersion(s);
        var u = getVersionFromCatalog(t, c, l);
        if (u) {
          return u;
        }
      }
    } else {
      return s;
    }
  }
  return null;
};

var hasPackageSupport = async (t, r) => {
  if (await getPackageVersion(t, r)) {
    return !0;
  }
  try {
    return !!o.createRequire(__dirname)(`${t}/package.json`)?.version;
  } catch (t) {
    return !1;
  }
};

var hasSvelteSupport = async t => hasPackageSupport("@gql.tada/svelte-support", t);

var hasVueSupport = async t => hasPackageSupport("@gql.tada/vue-support", t);

var jsonParse = async t => {
  var r = await u.readFile(t, "utf8");
  var i = s.parseJsonText(t, r);
  return s.convertToObject(i, []);
};

var isVSCodeInstalled = async () => {
  if (!process.env.HOME) {
    return !1;
  }
  var t = c.resolve(process.env.HOME, ".vscode");
  return !!await stat(t, xe.Directory);
};

var loadSuggestedExtensionsList = async t => {
  var r = process.cwd();
  var i = c.resolve(r, "/");
  while (r !== i) {
    if (await stat(c.resolve(r, ".git"), xe.Directory)) {
      break;
    } else if (await stat(c.resolve(r, ".vscode"), xe.Directory)) {
      break;
    }
    r = c.resolve(r, "..");
  }
  var n = c.resolve(r, ".vscode", "extensions.json");
  if (!await stat(n)) {
    return [];
  }
  var a;
  try {
    a = await jsonParse(n);
  } catch (t) {
    return [];
  }
  if (a && "object" == typeof a && "recommendations" in a) {
    return Array.isArray(a.recommendations) ? a.recommendations.filter((t => t && "string" == typeof t)).map((t => `${t}`.toLowerCase())) : [];
  } else {
    return [];
  }
};

var loadExtensionsList = async () => {
  if (!process.env.HOME) {
    return [];
  }
  var t = c.resolve(process.env.HOME, ".vscode");
  var r = c.resolve(t, "extensions", "extensions.json");
  if (!await stat(r)) {
    return [];
  }
  var i;
  try {
    i = await jsonParse(r);
  } catch (t) {
    return [];
  }
  return (Array.isArray(i) ? i : []).map((t => {
    if (!t || "object" != typeof t || !("identifier" in t)) {
      return null;
    }
    if (!t.identifier || "object" != typeof t.identifier || !("id" in t.identifier)) {
      return null;
    }
    return t.identifier.id && "string" == typeof t.identifier.id ? `${t.identifier.id}`.toLowerCase() : null;
  })).filter((t => !!t));
};

function emptyLine() {
  return text([ cmd(M.Style, L.BrightBlack), _.Vertical, "\n" ]);
}

function completedTask(t, r = !1) {
  return text([ emptyLine(), cmd(M.Style, L.BrightBlack), r ? _.BottomLeft : _.VerticalRight, " ", cmd(M.Style, L.Green), U.TickSwoosh, " ", cmd(M.Style, L.Foreground), t, "\n" ]);
}

function failedTask(t) {
  return text([ emptyLine(), cmd(M.Style, L.BrightBlack), _.BottomLeft, " ", cmd(M.Style, L.BrightRed), U.CrossSwoosh, " ", cmd(M.Style, L.Foreground), t, "\n" ]);
}

function warningTask(t) {
  return text([ emptyLine(), cmd(M.Style, L.BrightBlack), _.VerticalRight, " ", cmd(M.Style, L.BrightYellow), U.Warning, " ", cmd(M.Style, L.Foreground), t, "\n" ]);
}

function hintMessage(t) {
  return text([ cmd(M.Style, L.BrightBlack), `${_.VerticalRight} `, cmd(M.Style, L.BrightBlue), `${U.Info} `, cmd(M.Style, L.Blue), indent(t, text([ cmd(M.Style, L.BrightBlack), `${_.Vertical}   `, cmd(M.Style, L.Blue) ])) ]);
}

function runningTask(t) {
  var frame = r => text([ emptyLine(), cmd(M.Style, L.Magenta), H[r % H.length], " ", cmd(M.Style, L.Foreground), t.trim() ]);
  return concat([ fromValue(frame(0)), pipe(interval(150), map((t => frame(t + 1)))) ]);
}

function errorMessage(t) {
  return error([ "\n", cmd(M.Style, [ L.Red, L.Invert ]), ` ${U.Warning} Error `, cmd(M.Style, L.NoInvert), `\n${t.trim()}\n` ]);
}

var delay = (t = 700) => {
  if (process.env.CI) {
    return Promise.resolve();
  } else {
    return new Promise((r => {
      setTimeout(r, t);
    }));
  }
};

var Ce = function(t) {
  t.TITLE = "Doctor";
  t.DESCRIPTION = "Detects problems with your setup";
  t.CHECK_TS_VERSION = "Checking TypeScript version";
  t.CHECK_DEPENDENCIES = "Checking installed dependencies";
  t.CHECK_TSCONFIG = "Checking tsconfig.json";
  t.CHECK_EXTERNAL_FILES = "Checking external files support";
  t.CHECK_VSCODE = "Checking VSCode setup";
  t.CHECK_SCHEMA = "Checking schema";
  return t;
}(Ce || {});

async function* run() {
  yield function title(t, r) {
    var i = text([ cmd(M.Style, L.BrightBlack), _.TopLeft, " ", cmd(M.Style, [ L.Magenta, L.Invert ]), ` ${t.trim()} `, cmd(M.Style, [ L.NoInvert ]), "\n" ]);
    if (r) {
      i += text([ cmd(M.Style, L.BrightBlack), _.Vertical, ` ${r}\n` ]);
    }
    return i;
  }(Ce.TITLE, Ce.DESCRIPTION);
  yield runningTask(Ce.CHECK_TS_VERSION);
  await delay();
  var n;
  try {
    n = await (async () => {
      var r = t.resolve(process.cwd(), "package.json");
      var n = t.resolve(r);
      return JSON.parse(await i.readFile(n, "utf-8"));
    })();
  } catch (t) {
    yield failedTask(Ce.CHECK_TS_VERSION);
    throw errorMessage(`A ${code("package.json")} file was not found in the current working directory.\n` + hint("Try running the doctor command in your workspace folder."));
  }
  var a = await (async t => {
    var r = "typescript";
    var i = await getPackageVersion(r, t);
    if (i) {
      return i;
    }
    try {
      return (await import(r)).version || null;
    } catch (t) {
      return null;
    }
  })(n);
  if (!a) {
    yield failedTask(Ce.CHECK_TS_VERSION);
    throw errorMessage(`A version of ${code("typescript")} was not found in your dependencies.\n` + hint(`Is ${code("typescript")} installed in this package?`));
  } else if (!semverComply(a, ne.typescript)) {
    yield failedTask(Ce.CHECK_TS_VERSION);
    throw errorMessage(`The version of ${code("typescript")} in your dependencies is out of date.\n` + hint(`${code("gql.tada")} requires at least ${bold(ne.typescript)}`));
  }
  yield completedTask(Ce.CHECK_TS_VERSION);
  yield runningTask(Ce.CHECK_DEPENDENCIES);
  await delay();
  var s = semverComply(a, ne.typescript_embed_lsp);
  if (!s) {
    var l = await (async t => {
      var r = "@0no-co/graphqlsp";
      var i = await getPackageVersion(r, t);
      if (i) {
        return i;
      }
      try {
        return o.createRequire(__dirname)(`${r}/package.json`)?.version || null;
      } catch (t) {
        return null;
      }
    })(n);
    if (!l) {
      yield failedTask(Ce.CHECK_DEPENDENCIES);
      throw errorMessage(`A version of ${code("@0no-co/graphqlsp")} was not found in your dependencies.\n` + hint(`Is ${code("@0no-co/graphqlsp")} installed?`));
    } else if (!semverComply(l, ne.lsp)) {
      yield failedTask(Ce.CHECK_DEPENDENCIES);
      throw errorMessage(`The version of ${code("@0no-co/graphqlsp")} in your dependencies is out of date.\n` + hint(`${code("gql.tada")} requires at least ${bold(ne.lsp)}`));
    }
  }
  var c = await (async t => {
    var r = "gql.tada";
    var i = await getPackageVersion(r, t);
    if (i) {
      return i;
    }
    try {
      return o.createRequire(process.cwd())(`${r}/package.json`)?.version || null;
    } catch (t) {
      return null;
    }
  })(n);
  if (!c) {
    yield failedTask(Ce.CHECK_DEPENDENCIES);
    throw errorMessage(`A version of ${code("gql.tada")} was not found in your dependencies.\n` + hint(`Is ${code("gql.tada")} installed?`));
  } else if (!semverComply(c, "1.0.0")) {
    yield failedTask(Ce.CHECK_DEPENDENCIES);
    throw errorMessage(`The version of ${code("gql.tada")} in your dependencies is out of date.\n` + hint(`It's recommended to upgrade ${code("gql.tada")} to at least ${bold(ne.lsp)}`));
  }
  yield completedTask(Ce.CHECK_DEPENDENCIES);
  yield runningTask(Ce.CHECK_TSCONFIG);
  await delay();
  var u;
  try {
    u = await r.loadConfig();
  } catch (t) {
    yield failedTask(Ce.CHECK_TSCONFIG);
    throw externalError(`A ${code("tsconfig.json")} file was not found in the current working directory.`, t);
  }
  var d;
  try {
    d = r.parseConfig(u.pluginConfig, u.rootPath);
  } catch (t) {
    yield failedTask(Ce.CHECK_TSCONFIG);
    throw externalError(`The plugin configuration for ${code(s ? '"gql.tada/ts-plugin"' : '"@0no-co/graphqlsp"')} seems to be invalid.`, t);
  }
  yield completedTask(Ce.CHECK_TSCONFIG);
  yield* async function* runExternalFilesChecks(r, i) {
    var n = [];
    try {
      n = programFactory(r).createExternalFiles();
    } catch (t) {
      return;
    }
    if (n.length) {
      yield runningTask(Ce.CHECK_EXTERNAL_FILES);
      await delay();
      var a = new Set(n.map((r => t.extname(r.fileName))));
      if (a.has(".svelte") && !await hasSvelteSupport(i)) {
        yield failedTask(Ce.CHECK_EXTERNAL_FILES);
        throw errorMessage(`A version of ${code("@gql.tada/svelte-support")} must be installed for Svelte file support.\n` + hint(`Have you installed ${code("@gql.tada/svelte-support")}?`));
      }
      if (a.has(".vue") && !await hasVueSupport(i)) {
        yield failedTask(Ce.CHECK_EXTERNAL_FILES);
        throw errorMessage(`A version of ${code("@gql.tada/vue-support")} must be installed for Vue file support.\n` + hint(`Have you installed ${code("@gql.tada/vue-support")}?`));
      }
      yield completedTask(Ce.CHECK_EXTERNAL_FILES);
    }
  }(u, n);
  yield* async function* runVSCodeChecks() {
    var r = await loadSuggestedExtensionsList();
    var i = await isVSCodeInstalled();
    if (r.length || i) {
      yield runningTask(Ce.CHECK_VSCODE);
      await delay();
      var n = !1;
      var a = [];
      if (i) {
        if (!(a = await loadExtensionsList()).includes("graphql.vscode-graphql-syntax")) {
          if (!n) {
            n = !0;
            yield warningTask(Ce.CHECK_VSCODE);
          }
          yield hintMessage(`We recommend you to install the ${code('"GraphQL: Syntax Highlighting"')} extension for VSCode.\nSee: https://marketplace.visualstudio.com/items?itemName=GraphQL.vscode-graphql-syntax\n`);
        }
      }
      var o = a.includes("graphql.vscode-graphql") || r.includes("graphql.vscode-graphql");
      var s = await findGraphQLConfig();
      if (o && s) {
        if (!n) {
          n = !0;
          yield warningTask(Ce.CHECK_VSCODE);
        }
        var l = t.basename(s);
        yield hintMessage(`The ${code('"GraphQL: Language Feature Support"')} VSCode extension can cause problems!\nWhen enabled it may display invalid diagnostic errors for ${code("gql.tada")} code.\nCheck whether your ${code(l)} config only targets ${code(".graphql")} documents.\n`);
      }
      if (!n) {
        yield completedTask(Ce.CHECK_VSCODE);
      }
    }
  }();
  yield runningTask(Ce.CHECK_SCHEMA);
  await delay();
  try {
    await r.loadRef(d).load({
      rootPath: t.dirname(u.configPath)
    });
  } catch (t) {
    yield failedTask(Ce.CHECK_SCHEMA);
    throw externalError("Failed to load schema.", t);
  }
  yield completedTask(Ce.CHECK_SCHEMA, !0);
  await delay();
  yield function success() {
    return text([ "\n", cmd(M.Style, [ L.Green, L.Invert ]), " Done ", cmd(M.Style, L.NoInvert), D.Space, "You are all set and ready to go.\n" ]);
  }();
}

class DoctorCommand extends Command {
  static paths=[ [ "doctor" ] ];
  async execute() {
    var t = await initTTY().start(run());
    return exitCode() || ("object" == typeof t ? t.exit : 0);
  }
}

class InitCommand extends Command {
  static paths=[ [ "init" ] ];
  input=String$1({
    name: "dir"
  });
  async execute() {
    var {run: t} = await Promise.resolve().then((function() {
      return require("./runner-chunk.js");
    }));
    var r = c.resolve(process.cwd(), this.input);
    await t(r);
  }
}

var Oe = Object.assign((async function _main() {
  var t = new Cli({
    binaryVersion: process.env.npm_package_version || "0.0.0",
    binaryLabel: "gql.tada CLI",
    binaryName: "gql.tada"
  });
  t.register(CheckCommand);
  t.register(DoctorCommand);
  t.register(GenerateOutputCommand);
  t.register(GeneratePersisted);
  t.register(GenerateSchema);
  t.register(InitCommand);
  t.register(TurboCommand);
  await t.runExit(process.argv.slice(2));
}), ee);

exports.MINIMUM_VERSIONS = ne;

exports.generateOutput = generateOutput;

exports.generatePersisted = generatePersisted;

exports.generateSchema = generateSchema;

exports.generateTurbo = generateTurbo;

exports.main = Oe;

exports.programFactory = programFactory;

exports.semverComply = semverComply;
