// @gql.tada/cli-utils/dist/chunks/thread-chunk2.mjs
import e from "typescript";

import r from "node:v8";

import a from "node:vm";

import * as t from "node:path";

import { getSchemaNamesFromConfig as i } from "@gql.tada/internal";

import { findAllCallExpressions as o } from "@0no-co/graphqlsp/api";

import { e as n } from "./index-chunk2.mjs";

import { p as s } from "./index-chunk.mjs";

var l = 262144e4;

function findImportDeclaration(r) {
  var a = r;
  while (a) {
    if (e.isImportDeclaration(a)) {
      return a;
    }
    a = a.parent;
  }
  return;
}

function resolveModulePath(r, a, t) {
  var i = t.getCompilerOptions();
  var o = e.createCompilerHost(i);
  var n = e.resolveModuleName(r, a.fileName, i, o);
  if (n.resolvedModule) {
    return n.resolvedModule.resolvedFileName;
  }
  return;
}

function collectImportsFromSourceFile(r, a, i, o, n) {
  var s = [];
  var l = function getTadaOutputPaths(e) {
    var r = [];
    if ("schema" in e && e.tadaOutputLocation) {
      r.push(e.tadaOutputLocation);
    }
    return r;
  }(a);
  !function visit(a) {
    if (e.isImportDeclaration(a) && e.isStringLiteral(a.moduleSpecifier)) {
      var u = a.moduleSpecifier.text;
      if (!function isTadaImport(e, r, a) {
        if (e.startsWith(".")) {
          var i = t.dirname(r);
          var o = t.resolve(i, e);
          return a.some((e => {
            var r = t.resolve(e);
            return o === r || o.startsWith(r + t.sep);
          }));
        }
        return a.some((r => e === r || e.startsWith(r + "/")));
      }(u, r.fileName, l)) {
        var c = a.getFullText().trim();
        if (o) {
          var p = i(u, r.fileName, o);
          if (n) {
            if (p.endsWith(".ts") || p.endsWith(".tsx")) {
              p = p.replace(/\.ts$/, ".js").replace(/\.tsx$/, ".js");
            }
          } else {
            p = p.replace(/\.ts$/, "").replace(/\.tsx$/, "");
          }
          if (p && !p.includes("gql.tada")) {
            s.push({
              specifier: p,
              importClause: c.replace(u, p)
            });
          }
        } else {
          s.push({
            specifier: u,
            importClause: c
          });
        }
      }
    }
    e.forEachChild(a, visit);
  }(r);
  return s;
}

var u;

function forceGc() {
  if (void 0 === u) {
    var e = globalThis.gc;
    if ("function" == typeof e) {
      u = e;
    } else {
      try {
        r.setFlagsFromString("--expose-gc");
        u = a.runInNewContext("gc");
        r.setFlagsFromString("--no-expose-gc");
      } catch {
        u = null;
      }
    }
  }
  if (u) {
    u();
  }
}

var c = n((async function* _runTurbo(r) {
  var a = i(r.pluginConfig);
  var t = s(r);
  t.addSourceFile({
    fileId: "__gql-tada-override__.d.ts",
    sourceText: m,
    scriptKind: e.ScriptKind.TS
  });
  var n = t.createExternalFiles();
  if (n.length) {
    yield {
      kind: "EXTERNAL_WARNING"
    };
    await t.addVirtualFiles(n);
  }
  var u = t.build();
  var c = u.buildPluginInfo(r.pluginConfig);
  var f = u.getSourceFiles().map((e => e.fileName));
  yield {
    kind: "FILE_COUNT",
    fileCount: f.length
  };
  var d = new Map;
  var processSourceFile = (i, n, s, l) => {
    var u = i.fileName;
    var c = [];
    var m = [];
    var f = o(i, s, !1).nodes;
    var _loop = function(o) {
      var s = o.node.parent;
      if (!e.isCallExpression(s)) {
        return 0;
      }
      var f = n.getSourcePosition(i, s.getStart());
      u = f.fileName;
      if (!a.has(o.schema)) {
        m.push({
          message: o.schema ? `The '${o.schema}' schema is not in the configuration but was referenced by document.` : a.size > 1 ? "The document is not for a known schema. Have you re-generated the output file?" : "Multiple schemas are configured, but the document is not for a specific schema.",
          file: f.fileName,
          line: f.line,
          col: f.col
        });
        return 0;
      }
      var g = function traceCallToImportSource(r, a, t) {
        var i = t.getTypeChecker();
        var o = r.expression;
        var n;
        if (e.isIdentifier(o)) {
          n = o;
        } else if (e.isPropertyAccessExpression(o) && e.isIdentifier(o.expression)) {
          n = o.expression;
        }
        if (!n) {
          return;
        }
        var s = i.getSymbolAtLocation(n);
        if (!s || !s.declarations) {
          return;
        }
        for (var l of s.declarations) {
          var u = findImportDeclaration(l);
          if (u && e.isStringLiteral(u.moduleSpecifier)) {
            return resolveModulePath(u.moduleSpecifier.text, a, t);
          }
        }
        return;
      }(s, i, n.program);
      if (g && !d.has(g)) {
        var v = n.program.getSourceFile(g);
        if (v) {
          var h = Array.isArray(r.turboOutputPath) ? r.turboOutputPath.find((e => e.schemaName === o.schema))?.path : r.turboOutputPath;
          var y = collectImportsFromSourceFile(v, r.pluginConfig, t.resolveModuleName.bind(t), h, !!t.wasOriginallyNodeNext);
          d.set(g, {
            absolutePath: g,
            imports: y
          });
        }
      }
      var T = l.getTypeAtLocation(s);
      var F = l.getTypeAtLocation(o.node);
      if (!T.symbol || "TadaDocumentNode" !== T.symbol.getEscapedName()) {
        m.push({
          message: 'The discovered document is not of type "TadaDocumentNode".\nIf this is unexpected, please file an issue describing your case.',
          file: f.fileName,
          line: f.line,
          col: f.col
        });
        return 0;
      }
      var S = "value" in F && "string" == typeof F.value && !(F.flags & e.TypeFlags.StringLiteral) ? JSON.stringify(F.value) : l.typeToString(F, s, p);
      var N = l.typeToString(T, s, p);
      c.push({
        schemaName: o.schema,
        argumentKey: S,
        documentType: N
      });
    };
    for (var g of f) {
      if (0 === _loop(g)) {
        continue;
      }
    }
    return {
      filePath: u,
      documents: c,
      warnings: m
    };
  };
  var isHeapOverSoftLimit = () => {
    if (process.memoryUsage().heapUsed < l) {
      return !1;
    }
    forceGc();
    return process.memoryUsage().heapUsed >= l;
  };
  var g = u.program.getTypeChecker();
  var v = 0;
  for (var h of f) {
    if (v > 0 && (v >= 1e3 || isHeapOverSoftLimit())) {
      c = (u = t.build()).buildPluginInfo(r.pluginConfig);
      forceGc();
      g = u.program.getTypeChecker();
      v = 0;
    }
    var y = u.getSourceFile(h);
    if (!y) {
      continue;
    }
    var {filePath: T, documents: F, warnings: S} = processSourceFile(y, u, c, g);
    v++;
    yield {
      kind: "FILE_TURBO",
      filePath: T,
      documents: F,
      warnings: S
    };
  }
  if (d.size > 0) {
    yield {
      kind: "GRAPHQL_SOURCES",
      sources: Array.from(d.values())
    };
  }
}));

var p = e.TypeFormatFlags.NoTruncation | e.TypeFormatFlags.NoTypeReduction | e.TypeFormatFlags.InTypeAlias | e.TypeFormatFlags.UseFullyQualifiedType | e.TypeFormatFlags.GenerateNamesForShadowedTypeParams | e.TypeFormatFlags.UseAliasDefinedOutsideCurrentScope | e.TypeFormatFlags.AllowUniqueESSymbolType | e.TypeFormatFlags.WriteTypeArgumentsOfSignature;

var m = "\nimport * as _gqlTada from 'gql.tada';\ndeclare module 'gql.tada' {\n  interface setupCache {\n    readonly __cacheDisabled: true;\n  }\n}\n".trim();

export { c as runTurbo };
