// @gql.tada/cli-utils/dist/chunks/thread-chunk2.js
var e = require("typescript");

var r = require("node:v8");

var a = require("node:vm");

var t = require("node:path");

var i = require("@gql.tada/internal");

var o = require("@0no-co/graphqlsp/api");

var n = require("./index-chunk2.js");

var s = require("./index-chunk.js");

function _interopNamespaceDefault(e) {
  var r = Object.create(null);
  if (e) {
    Object.keys(e).forEach((function(a) {
      if ("default" !== a) {
        var t = Object.getOwnPropertyDescriptor(e, a);
        Object.defineProperty(r, a, t.get ? t : {
          enumerable: !0,
          get: function() {
            return e[a];
          }
        });
      }
    }));
  }
  r.default = e;
  return r;
}

var l = _interopNamespaceDefault(t);

var u = 262144e4;

function findImportDeclaration(r) {
  var a = r;
  while (a) {
    if (e.isImportDeclaration(a)) {
      return a;
    }
    a = a.parent;
  }
  return;
}

function resolveModulePath(r, a, t) {
  var i = t.getCompilerOptions();
  var o = e.createCompilerHost(i);
  var n = e.resolveModuleName(r, a.fileName, i, o);
  if (n.resolvedModule) {
    return n.resolvedModule.resolvedFileName;
  }
  return;
}

function collectImportsFromSourceFile(r, a, t, i, o) {
  var n = [];
  var s = function getTadaOutputPaths(e) {
    var r = [];
    if ("schema" in e && e.tadaOutputLocation) {
      r.push(e.tadaOutputLocation);
    }
    return r;
  }(a);
  !function visit(a) {
    if (e.isImportDeclaration(a) && e.isStringLiteral(a.moduleSpecifier)) {
      var u = a.moduleSpecifier.text;
      if (!function isTadaImport(e, r, a) {
        if (e.startsWith(".")) {
          var t = l.dirname(r);
          var i = l.resolve(t, e);
          return a.some((e => {
            var r = l.resolve(e);
            return i === r || i.startsWith(r + l.sep);
          }));
        }
        return a.some((r => e === r || e.startsWith(r + "/")));
      }(u, r.fileName, s)) {
        var c = a.getFullText().trim();
        if (i) {
          var p = t(u, r.fileName, i);
          if (o) {
            if (p.endsWith(".ts") || p.endsWith(".tsx")) {
              p = p.replace(/\.ts$/, ".js").replace(/\.tsx$/, ".js");
            }
          } else {
            p = p.replace(/\.ts$/, "").replace(/\.tsx$/, "");
          }
          if (p && !p.includes("gql.tada")) {
            n.push({
              specifier: p,
              importClause: c.replace(u, p)
            });
          }
        } else {
          n.push({
            specifier: u,
            importClause: c
          });
        }
      }
    }
    e.forEachChild(a, visit);
  }(r);
  return n;
}

var c;

function forceGc() {
  if (void 0 === c) {
    var e = globalThis.gc;
    if ("function" == typeof e) {
      c = e;
    } else {
      try {
        r.setFlagsFromString("--expose-gc");
        c = a.runInNewContext("gc");
        r.setFlagsFromString("--no-expose-gc");
      } catch {
        c = null;
      }
    }
  }
  if (c) {
    c();
  }
}

var p = n.expose((async function* _runTurbo(r) {
  var a = i.getSchemaNamesFromConfig(r.pluginConfig);
  var t = s.programFactory(r);
  t.addSourceFile({
    fileId: "__gql-tada-override__.d.ts",
    sourceText: d,
    scriptKind: e.ScriptKind.TS
  });
  var n = t.createExternalFiles();
  if (n.length) {
    yield {
      kind: "EXTERNAL_WARNING"
    };
    await t.addVirtualFiles(n);
  }
  var l = t.build();
  var c = l.buildPluginInfo(r.pluginConfig);
  var p = l.getSourceFiles().map((e => e.fileName));
  yield {
    kind: "FILE_COUNT",
    fileCount: p.length
  };
  var m = new Map;
  var processSourceFile = (i, n, s, l) => {
    var u = i.fileName;
    var c = [];
    var p = [];
    var d = o.findAllCallExpressions(i, s, !1).nodes;
    var _loop = function(o) {
      var s = o.node.parent;
      if (!e.isCallExpression(s)) {
        return 0;
      }
      var d = n.getSourcePosition(i, s.getStart());
      u = d.fileName;
      if (!a.has(o.schema)) {
        p.push({
          message: o.schema ? `The '${o.schema}' schema is not in the configuration but was referenced by document.` : a.size > 1 ? "The document is not for a known schema. Have you re-generated the output file?" : "Multiple schemas are configured, but the document is not for a specific schema.",
          file: d.fileName,
          line: d.line,
          col: d.col
        });
        return 0;
      }
      var v = function traceCallToImportSource(r, a, t) {
        var i = t.getTypeChecker();
        var o = r.expression;
        var n;
        if (e.isIdentifier(o)) {
          n = o;
        } else if (e.isPropertyAccessExpression(o) && e.isIdentifier(o.expression)) {
          n = o.expression;
        }
        if (!n) {
          return;
        }
        var s = i.getSymbolAtLocation(n);
        if (!s || !s.declarations) {
          return;
        }
        for (var l of s.declarations) {
          var u = findImportDeclaration(l);
          if (u && e.isStringLiteral(u.moduleSpecifier)) {
            return resolveModulePath(u.moduleSpecifier.text, a, t);
          }
        }
        return;
      }(s, i, n.program);
      if (v && !m.has(v)) {
        var g = n.program.getSourceFile(v);
        if (g) {
          var h = Array.isArray(r.turboOutputPath) ? r.turboOutputPath.find((e => e.schemaName === o.schema))?.path : r.turboOutputPath;
          var y = collectImportsFromSourceFile(g, r.pluginConfig, t.resolveModuleName.bind(t), h, !!t.wasOriginallyNodeNext);
          m.set(v, {
            absolutePath: v,
            imports: y
          });
        }
      }
      var T = l.getTypeAtLocation(s);
      var F = l.getTypeAtLocation(o.node);
      if (!T.symbol || "TadaDocumentNode" !== T.symbol.getEscapedName()) {
        p.push({
          message: 'The discovered document is not of type "TadaDocumentNode".\nIf this is unexpected, please file an issue describing your case.',
          file: d.fileName,
          line: d.line,
          col: d.col
        });
        return 0;
      }
      var S = "value" in F && "string" == typeof F.value && !(F.flags & e.TypeFlags.StringLiteral) ? JSON.stringify(F.value) : l.typeToString(F, s, f);
      var N = l.typeToString(T, s, f);
      c.push({
        schemaName: o.schema,
        argumentKey: S,
        documentType: N
      });
    };
    for (var v of d) {
      if (0 === _loop(v)) {
        continue;
      }
    }
    return {
      filePath: u,
      documents: c,
      warnings: p
    };
  };
  var isHeapOverSoftLimit = () => {
    if (process.memoryUsage().heapUsed < u) {
      return !1;
    }
    forceGc();
    return process.memoryUsage().heapUsed >= u;
  };
  var v = l.program.getTypeChecker();
  var g = 0;
  for (var h of p) {
    if (g > 0 && (g >= 1e3 || isHeapOverSoftLimit())) {
      c = (l = t.build()).buildPluginInfo(r.pluginConfig);
      forceGc();
      v = l.program.getTypeChecker();
      g = 0;
    }
    var y = l.getSourceFile(h);
    if (!y) {
      continue;
    }
    var {filePath: T, documents: F, warnings: S} = processSourceFile(y, l, c, v);
    g++;
    yield {
      kind: "FILE_TURBO",
      filePath: T,
      documents: F,
      warnings: S
    };
  }
  if (m.size > 0) {
    yield {
      kind: "GRAPHQL_SOURCES",
      sources: Array.from(m.values())
    };
  }
}));

var f = e.TypeFormatFlags.NoTruncation | e.TypeFormatFlags.NoTypeReduction | e.TypeFormatFlags.InTypeAlias | e.TypeFormatFlags.UseFullyQualifiedType | e.TypeFormatFlags.GenerateNamesForShadowedTypeParams | e.TypeFormatFlags.UseAliasDefinedOutsideCurrentScope | e.TypeFormatFlags.AllowUniqueESSymbolType | e.TypeFormatFlags.WriteTypeArgumentsOfSignature;

var d = "\nimport * as _gqlTada from 'gql.tada';\ndeclare module 'gql.tada' {\n  interface setupCache {\n    readonly __cacheDisabled: true;\n  }\n}\n".trim();

exports.runTurbo = p;
